### FIG5
##### pgdlnc基因m6A修饰
```r
#REPIC数据库下载 m6A peak位点（人类hg38，小鼠mm10）
wget https://repicmod.uchicago.edu/repic/data/download/m6A=sites=species=human=hg38.txt.gz --no-check-certificate
wget https://repicmod.uchicago.edu/repic/data/download/m6A=sites=species=mouse=mm10.txt.gz --no-check-certificate
gzip -d "/home/zhzhang/PG/m6a/m6A=sites=species=human=hg38.txt.gz"
gzip -d "/home/zhzhang/PG/m6a/m6A=sites=species=mouse=mm10.txt.gz"



#提取简约信息（POS）,转化为非冗余的m6A位点bed（所有潜在的m6A位点）
#人
tail -n +2 "/home/zhzhang/PG/m6a/m6A=sites=species=human=hg38.txt"|awk '{print $1}' > /home/zhzhang/PG/m6a/m6A=sites=human=hg38.txt
grep -w "+" /home/zhzhang/PG/m6a/m6A=sites=human=hg38.txt|sed 's/\[+\]//g'|awk -F ":" '{print $1"\t"$2}'|awk -F "-" '{print $1"\t"$2}'|sed 's/^chr//g'|bedtools sort|bedtools merge|awk '{print $0"\t+"}' > /home/zhzhang/PG/m6a/Homo_sapiens.m6Asites.bed
grep -w "-" /home/zhzhang/PG/m6a/m6A=sites=human=hg38.txt|sed 's/\[-\]//g'|awk -F ":" '{print $1"\t"$2}'|awk -F "-" '{print $1"\t"$2}'|sed 's/^chr//g'|bedtools sort|bedtools merge|awk '{print $0"\t-"}' >> /home/zhzhang/PG/m6a/Homo_sapiens.m6Asites.bed
cat /home/zhzhang/PG/m6a/Homo_sapiens.m6Asites.bed|bedtools sort|awk '{print $1"\t"$2"\t"$3"\tm6Asite_"NR"\tm6Asite\t"$4}' > /home/zhzhang/PG/m6a/Homo_sapiens.m6Asites.sort.bed
rm /home/zhzhang/PG/m6a/Homo_sapiens.m6Asites.bed
#小鼠(外加liftover到grch39)
tail -n +2 "/home/zhzhang/PG/m6a/m6A=sites=species=mouse=mm10.txt"|awk '{print $1}' > /home/zhzhang/PG/m6a/m6A=sites=mouse=mm10.txt
grep -w "+" "/home/zhzhang/PG/m6a/m6A=sites=mouse=mm10.txt"|sed 's/\[+\]//g'|awk -F ":" '{print $1"\t"$2}'|awk -F "-" '{print $1"\t"$2}'|bedtools sort|bedtools merge|awk '{print $0"\t+"}' > /home/zhzhang/PG/m6a/Mus_musculus.m6Asites.bed
grep -w "-" "/home/zhzhang/PG/m6a/m6A=sites=mouse=mm10.txt"|sed 's/\[-\]//g'|awk -F ":" '{print $1"\t"$2}'|awk -F "-" '{print $1"\t"$2}'|bedtools sort|bedtools merge|awk '{print $0"\t-"}' >> /home/zhzhang/PG/m6a/Mus_musculus.m6Asites.bed
cat /home/zhzhang/PG/m6a/Mus_musculus.m6Asites.bed|bedtools sort|awk '{print $1"\t"$2"\t"$3"\tm6Asite_"NR"\tm6Asite\t"$4}' > /home/zhzhang/PG/m6a/Mus_musculus.m6Asites.sort.mm10.bed
rm /home/zhzhang/PG/m6a/Mus_musculus.m6Asites.bed
liftOver /home/zhzhang/PG/m6a/Mus_musculus.m6Asites.sort.mm10.bed /home/zhzhang/software/liftover/mm10ToMm39.over.chain.gz /home/zhzhang/PG/m6a/Mus_musculus.m6Asites.sort.mm39.bed /home/zhzhang/PG/m6a/Mus_musculus.unmap
rm /home/zhzhang/PG/m6a/Mus_musculus.unmap
cat /home/zhzhang/PG/m6a/Mus_musculus.m6Asites.sort.mm39.bed |sed 's/^chr//g'|bedtools sort > /home/zhzhang/PG/m6a/Mus_musculus.m6Asites.sort.bed
rm /home/zhzhang/PG/m6a/Mus_musculus.m6Asites.sort.mm*.bed



```
```r
#非冗余的m6a位点与lncRNA基因外显子交集，获取每个基因外显子具有的m6a位点数量
#人
bedtools intersect -a /home/zhzhang/PG/RBP/Homo_sapiens.allgeneexon.bed -b /home/zhzhang/PG/m6a/Homo_sapiens.m6Asites.sort.bed -c -s > /home/zhzhang/PG/m6a/Homo_sapiens.allgeneexon_intersect_m6A.txt
#小鼠
bedtools intersect -a /home/zhzhang/PG/RBP/Mus_musculus.allgeneexon.bed -b /home/zhzhang/PG/m6a/Mus_musculus.m6Asites.sort.bed -c -s > /home/zhzhang/PG/m6a/Mus_musculus.allgeneexon_intersect_m6A.txt



#m6a位点与基因间区交集
bedtools intersect -a /home/zhzhang/PG/Evolution/conserve/Homo_sapiens.20000random_3kb_intergenic.bed -b /home/zhzhang/PG/m6a/Homo_sapiens.m6Asites.sort.bed -c -s > /home/zhzhang/PG/m6a/Homo_sapiens.intergenic_intersect_m6A.txt
bedtools intersect -a /home/zhzhang/PG/Evolution/conserve/Mus_musculus.20000random_3kb_intergenic.bed -b /home/zhzhang/PG/m6a/Mus_musculus.m6Asites.sort.bed -c -s > /home/zhzhang/PG/m6a/Mus_musculus.intergenic_intersect_m6A.txt


```
##### 对比蛋白编码/lncRNA基因转录本以及基因间区的m6a位点密度
```r
#函数统计每个基因全部外显子&基因间区具有的m6a位点密度
#a输入每个基因全部外显子与m6a位点交集文件，b输入每个基因间区与m6a位点交集文件
#c输入基因分类文件，e输入全部基因外显子bed文件
geneexonM6Ad <- function(a,b,c,e){
  #导入全部外显子与m6a位点交集,获取每个基因全部外显子(全部)中每种m6a位点数量
  allgeneexon_intersect_m6a <- data.table::fread(a, header=FALSE)%>%
    data.frame()%>%
    select(4,7)%>%
    group_by(V4)%>%
    summarise(num=sum(V7))
  colnames(allgeneexon_intersect_m6a) <- c("geneid","num")
  #计算基因全部外显子（全部转录本）长度
  allgeneexon <- read.delim(e, header=FALSE)%>%
    mutate(len=(V3-V2+1)/1000)%>%
    group_by(V4)%>%
    summarise(len=sum(len))
  colnames(allgeneexon) <- c("geneid","len")
  #合并长度和m6a位点数量信息，计算密度（m6a位点数量/基因全部外显子长度）
  allgeneexon_m6a_len <- left_join(allgeneexon_intersect_m6a,allgeneexon,by="geneid")%>%
    mutate(m6adensity=num/len)
  #导入基因分类
  geneid_class <- read.delim(c)
  allgeneexon_m6a_len <- left_join(geneid_class,allgeneexon_m6a_len,by="geneid")%>%
    filter(type!="Interference lncRNA")
  #导入基因间区与RBP交集,获取每个基因间区每种RBP的结合长度比例
  intergenic_intersect_m6a <- data.table::fread(b, header=FALSE)%>%
    data.frame()%>%
    select(4,7)%>%
    group_by(V4)%>%
    summarise(num=sum(V7))%>%
    mutate(len=3,type="Random intergenic")%>%
    mutate(m6adensity=num/len)%>%
    select(1,4,2,3,5)
  colnames(intergenic_intersect_m6a) <- colnames(allgeneexon_m6a_len)
  #合并
  allm6ainfo <- rbind(allgeneexon_m6a_len,intergenic_intersect_m6a)
}

#人
hsm6ad <- geneexonM6Ad(a="/home/zhzhang/PG/m6a/Homo_sapiens.allgeneexon_intersect_m6A.txt",
            b="/home/zhzhang/PG/m6a/Homo_sapiens.intergenic_intersect_m6A.txt",
            c="/home/zhzhang/PG/RNAseq/Homo_sapiens/Homo_sapiens.geneid_class.txt",
            e="/home/zhzhang/PG/RBP/Homo_sapiens.allgeneexon.bed")
#小鼠
mmm6ad <- geneexonM6Ad(a="/home/zhzhang/PG/m6a/Mus_musculus.allgeneexon_intersect_m6A.txt",
            b="/home/zhzhang/PG/m6a/Mus_musculus.intergenic_intersect_m6A.txt",
            c="/home/zhzhang/PG/RNAseq/Mus_musculus/Mus_musculus.geneid_class.txt",
            e="/home/zhzhang/PG/RBP/Mus_musculus.allgeneexon.bed")
#合并所有物种结果
ALLSP_m6a <- rbind(mutate(hsm6ad,sp="Human"),mutate(mmm6ad,sp="Mouse"))
ALLSP_m6a$type <- factor(ALLSP_m6a$type,levels = c("Protein-coding","Pseudogene-derived lncRNA",
                                                           "Non-pseudogene-derived lncRNA","Random intergenic"))
data.table::fwrite(ALLSP_m6a,file ="/home/zhzhang/PG/m6a/SPhsmm_3gene_m6Adensity.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#统计
tj <- group_by(ALLSP_m6a,sp,type)%>%
  summarise(mean=mean(m6adensity),median=median(m6adensity))
data.table::fwrite(tj,file ="/home/zhzhang/PG/m6a/SPhsmm_3gene_m6Adensity.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#plot
phs <- ggplot(data=filter(ALLSP_m6a,sp=="Human"), aes(x=type,y=m6adensity))+
  geom_boxplot(fatten = 3,outlier.alpha = 0,width=0.5,notch=T,aes(fill=type))+
  geom_signif(map_signif_level=T,y_position=c(7,3.5,5),tip_length = 0.01,
              comparisons = list(c("Protein-coding","Pseudogene-derived lncRNA"),
                                 c("Non-pseudogene-derived lncRNA","Pseudogene-derived lncRNA"),
                                 c("Pseudogene-derived lncRNA","Random intergenic")
              ))+
  scale_fill_manual(values=c("#BB5A5D","#7197AD","#FAA465","#8491B4"),
                    limits=c("Protein-coding","Pseudogene-derived lncRNA",
                             "Non-pseudogene-derived lncRNA","Random intergenic"))+
  theme_half_open()+
  coord_cartesian(ylim = c(0, 8.5))+
  scale_y_continuous(breaks = c(0,1,2,3,4,5,6,7,8))+
  scale_x_discrete(labels = c("Protein-coding","Pseudogene-\nderived lncRNA",
                              "Non-pseudogene-\nderived lncRNA","Random\nintergenic"))+
  labs(y = "Density of m6A sites",x =NULL,fill = NULL,color = NULL)+
  theme(axis.title = element_text(size = 20),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),legend.position = "none") +
  theme(axis.text.x = element_text(angle =40)) +
  theme(axis.text.x = element_text(hjust=1,vjust = 1))
ggsave("/home/zhzhang/PG/m6a/Homo_sapiens.3gene_m6Adensity.png",
       phs,width = 4.5, height = 5,dpi=1200, units = "in", device='png',bg = "transparent")

pmm <- ggplot(data=filter(ALLSP_m6a,sp=="Mouse"), aes(x=type,y=m6adensity))+
  geom_boxplot(fatten = 3,outlier.alpha = 0,width=0.5,notch=T,aes(fill=type))+
  geom_signif(map_signif_level=T,y_position=c(7.5,2.5,4.5),tip_length = 0.01,
              comparisons = list(c("Protein-coding","Pseudogene-derived lncRNA"),
                                 c("Non-pseudogene-derived lncRNA","Pseudogene-derived lncRNA"),
                                 c("Pseudogene-derived lncRNA","Random intergenic")
              ))+
  scale_fill_manual(values=c("#BB5A5D","#7197AD","#FAA465","#8491B4"),
                    limits=c("Protein-coding","Pseudogene-derived lncRNA",
                             "Non-pseudogene-derived lncRNA","Random intergenic"))+
  theme_half_open()+
  coord_cartesian(ylim = c(0, 9.5))+
  scale_y_continuous(breaks = c(0,1,2,3,4,5,6,7,8))+
  scale_x_discrete(labels = c("Protein-coding","Pseudogene-\nderived lncRNA",
                              "Non-pseudogene-\nderived lncRNA","Random\nintergenic"))+
  labs(y = "Density of m6A sites",x =NULL,fill = NULL,color = NULL)+
  theme(axis.title = element_text(size = 20),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),legend.position = "none") +
  theme(axis.text.x = element_text(angle =40)) +
  theme(axis.text.x = element_text(hjust=1,vjust = 1))
ggsave("/home/zhzhang/PG/m6a/Mus_musculus.3gene_m6Adensity.png",
       pmm,width = 4.5, height = 5,dpi=1200, units = "in", device='png',bg = "transparent")


```


##### pgdlnc基因RBP结合
```r
#POSTAR3数据库 实验检测的RBP结合位点(HG38,MM10)
wget https://cloud.tsinghua.edu.cn/seafhttp/files/d05e6a97-0c75-4029-86a8-8b135368aae4/human.txt.gz
gzip -d "/home/zhzhang/PG/RBP/human.txt.gz"
#转bed【chr start end RBPname ID strand】(去掉RBP不明确的条目)
#人
grep -w -v RBP_occupancy /home/zhzhang/PG/RBP/human.txt|awk '{print $1"\t"$2"\t"$3"\t"$6"\t"$4"\t"$5}'|sed 's/^chr//g'|bedtools sort > /home/zhzhang/PG/RBP/Homo_sapiens.RBP_BS.bed

```
```r
#RBP结合位点信息整合(获得每个RBP非冗余的结合位点，再合并到一起)（所有RBP所有潜在的结合位点）
#人(220种)
cat /home/zhzhang/PG/RBP/Homo_sapiens.RBP_BS.bed|awk '{print $4}'|sort|uniq > /home/zhzhang/PG/RBP/Homo_sapiens.RBPname.txt
################"/home/zhzhang/PG/RBP/hsRBPrmd.sh"内容
#!/bin/bash
cat /home/zhzhang/PG/RBP/Homo_sapiens.RBPname.txt|while read i
do
RBP=${i}
grep -w "${RBP}" /home/zhzhang/PG/RBP/Homo_sapiens.RBP_BS.bed|awk '$6=="+"{print $0}'|bedtools sort|bedtools merge|awk -v RBP="${RBP}" '{print $1"\t"$2"\t"$3"\t"RBP"\tRBPBS\t+"}' > /home/zhzhang/PG/RBP/hstmp/${RBP}.bed
grep -w "${RBP}" /home/zhzhang/PG/RBP/Homo_sapiens.RBP_BS.bed|awk '$6=="-"{print $0}'|bedtools sort|bedtools merge|awk -v RBP="${RBP}" '{print $1"\t"$2"\t"$3"\t"RBP"\tRBPBS\t-"}' >> /home/zhzhang/PG/RBP/hstmp/${RBP}.bed
cat /home/zhzhang/PG/RBP/hstmp/${RBP}.bed >> /home/zhzhang/PG/RBP/hstmp/Homo_sapiens.RBP_BS.NonRedundant.bed
echo "${RBP}" >> /home/zhzhang/PG/RBP/hstmp/Homo_sapiens.allRBP.log
done
bedtools sort -i /home/zhzhang/PG/RBP/hstmp/Homo_sapiens.RBP_BS.NonRedundant.bed > /home/zhzhang/PG/RBP/Homo_sapiens.RBP_BS.NonRedundant.sort.bed
###############################################################
nohup "/home/zhzhang/PG/RBP/hsRBPrmd.sh" &



#非冗余的RBP结合位点与基因外显子交集，获取每个基因外显子具有的RBP情况
#人
#提取所有基因外显子bed【chr start end geneid tid strand】
cat "/home/zhzhang/PG/RNAseqdata/newGTF/Homo_sapiens.GRCh38.108.chr.rmpg.novellncRNA.gtf"|awk '$3=="exon" && $13=="transcript_id"{print $1"\t"$4"\t"$5"\t"$10"\t"$14"\t"$7} $3=="exon" && $11=="transcript_id"{print $1"\t"$4"\t"$5"\t"$10"\t"$12"\t"$7}'|sed "s/\"//g;s/\;//g"|bedtools sort > /home/zhzhang/PG/RBP/Homo_sapiens.allgeneexon.bed
#交集
bedtools intersect -a /home/zhzhang/PG/RBP/Homo_sapiens.allgeneexon.bed -b /home/zhzhang/PG/RBP/Homo_sapiens.RBP_BS.NonRedundant.sort.bed -wo -s > /home/zhzhang/PG/RBP/Homo_sapiens.allgeneexon_intersect_RBP.txt



```
##### 统计每个基因全部外显子(全部转录本)具有的RBP种类及对应的结合比例(结合长度/全部外显子长度)信息文件
```r
#函数统计每个基因全部外显子&基因间区具有的RBP种类及对应结合长度比例的信息文件
#a输入每个基因全部外显子与RBP交集文件，e输入全部基因外显子bed文件
#c输出每个基因外显子区域&基因间区拥有的RBP种类及其结合长度比例统计结果
geneexonRBP <- function(a,c,e){
  #导入启动子与RBP交集,获取每个基因全部外显子(全部)中每种RBP的结合长度
  allgeneexon_intersect_RBP <- data.table::fread(a, header=FALSE)%>%
    data.frame()%>%
    select(4,10,13)%>%
    separate(V10,c("RBP"),remove = T,sep = "_")%>%
    group_by(V4,RBP)%>%
    summarise(len=sum(V13))
  colnames(allgeneexon_intersect_RBP) <- c("geneid","RBP","num")
  #计算基因全部外显子（全部转录本）长度
  allgeneexon <- read.delim(e, header=FALSE)%>%
    mutate(len=(V3-V2+1)/1000)%>%
    group_by(V4)%>%
    summarise(len=sum(len))
  colnames(allgeneexon) <- c("geneid","len")
  #合并长度和RBP信息，计算结合长度占比（RBP总结合长度/基因全部外显子长度）
  allgeneexon_RBP_len <- left_join(allgeneexon_intersect_RBP,allgeneexon,by="geneid")%>%
    mutate(num=num/1000)%>%
    mutate(ratio=num/len)%>%
    select(1,2,5)
  allrbpinfo <- allgeneexon_RBP_len
  #储存
  data.table::fwrite(allrbpinfo,file =c,sep = '\t',row.names = F,quote = F,col.names = T)
}

#人
geneexonRBP(a="/home/zhzhang/PG/RBP/Homo_sapiens.allgeneexon_intersect_RBP.txt",
            c="/home/zhzhang/PG/RBP/Homo_sapiens.allgeneexonAintergenic.RBP.txt",
            e="/home/zhzhang/PG/RBP/Homo_sapiens.allgeneexon.bed")


```
##### 对比蛋白编码/lncRNA基因转录本的RBP种类数量
```r
#函数根据基因具有的RBP种类及其结合比例信息文件，统计RBP种类数
#a输入基因具有的RBP种类及其结合比例信息文件，b输入物种英文名，c输入物种基因分类文件，d输入基因间区文件
getRBP_NUM <- function(a,b,c,d){
  #导入基因具有的RBP种类及其结合比例信息文件,计算RBP种类数
  allgene_RBP <- read.delim(a)%>%
    mutate(ratio=1)%>%
    group_by(geneid)%>%
    summarise(RBPnum=sum(ratio))
  #导入基因分类
  geneid_class <- read.delim(c)
  #导入基因间区分类
  intergenic <- read.delim(d, header=FALSE)%>%
    select(1)%>%
    mutate(type="Random intergenic")
  colnames(intergenic)[1] <- "geneid"
  #合并
  allclass <- rbind(geneid_class,intergenic)
  geneid_class_RBP <- left_join(allclass,allgene_RBP,by="geneid")%>%
    filter(type!="Interference lncRNA")%>%
    mutate(sp=b)
  geneid_class_RBP$RBPnum[is.na(geneid_class_RBP$RBPnum)==T] <- 0
  return(geneid_class_RBP)
}

#人
hs_RBP_NUM <- getRBP_NUM(a="/home/zhzhang/PG/RBP/Homo_sapiens.allgeneexonAintergenic.RBP.txt",
                         b="Human",
                         c="~/PG/RNAseq/Homo_sapiens/Homo_sapiens.geneid_class.txt",
                         d="~/PG/Evolution/conserve/Homo_sapiens.20000random_3kb_intergenic.phastCons.txt")
#合并所有物种结果
ALLSP_RBP_NUM <- hs_RBP_NUM
ALLSP_RBP_NUM$type <- factor(ALLSP_RBP_NUM$type,levels = c("Protein-coding","Pseudogene-derived lncRNA",
                                                             "Non-pseudogene-derived lncRNA","Random intergenic"))
data.table::fwrite(ALLSP_RBP_NUM,file ="/home/zhzhang/PG/RBP/SPhsmm_3gene_RBPrichness.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#统计具有RBP结合的基因比例
ALLSP_RBP_NUM_forratio <- ALLSP_RBP_NUM
ALLSP_RBP_NUM_forratio$RBPnum[ALLSP_RBP_NUM_forratio$RBPnum >0] <- 1
tjratio <- group_by(filter(ALLSP_RBP_NUM_forratio,type!="Random intergenic"),sp,type)%>%
  summarise(allnum=n(),ownRBPnum=sum(RBPnum))%>%
  mutate(noRBPnum=allnum-ownRBPnum,ratio=ownRBPnum*100/allnum)
data.table::fwrite(tjratio,file ="/home/zhzhang/PG/RBP/SPhsmm_3gene_ownRBPgeneratio.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#检验人类两类lncRNA的拥有RBP结合的基因比例之间是否有显著性差异(p-value = 0.6268)
fisher.test(tjratio[c(1,3),c(4,5)])
#plot具有RBP结合的基因的比例
tjratio$type <- factor(tjratio$type,levels = c("Protein-coding","Pseudogene-derived lncRNA",
                                           "Non-pseudogene-derived lncRNA"))
ownRBPratio <- ggplot(data=filter(tjratio,sp=="Human"),
                      aes(x=type,y=ratio))+
  geom_col(width=0.5,aes(fill=type))+
  geom_signif(annotations="N.S.",y_position=c(25),tip_length = 0.01,
              xmin = 2,xmax = 3)+
  scale_fill_manual(values=c("#BB5A5D","#7197AD","#FAA465","#8491B4"),
                    limits=c("Protein-coding","Pseudogene-derived lncRNA",
                             "Non-pseudogene-derived lncRNA"))+
  theme_half_open()+
  scale_y_continuous(limits = c(0,90),breaks = c(0,20,40,60,80))+
  scale_x_discrete(labels = c("Protein-coding","Pseudogene-\nderived lncRNA",
                              "Non-pseudogene-\nderived lncRNA","Random\nintergenic"))+
  labs(y = "Proportion of genes\nwith RBP binding sites",
       x =NULL,fill = NULL,color = NULL)+
  theme(axis.title = element_text(size = 16),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),legend.position = "none") +
  theme(axis.text.x = element_text(angle =30)) +
  theme(axis.text.x = element_text(hjust=1,vjust = 1))
ggsave("/home/zhzhang/PG/RBP/Homo_sapiens.3gene_ownRBPgeneratio.png",
       ownRBPratio,width = 5, height = 5,dpi=1200, units = "in", device='png',bg = "transparent")

#统计RBP丰富度(排除基因间区，以及未被检测到与RBP结合的基因)
tj <- group_by(filter(ALLSP_RBP_NUM,type!="Random intergenic" & RBPnum!=0),sp,type)%>%
  summarise(mean=mean(RBPnum),median=median(RBPnum))
data.table::fwrite(tj,file ="/home/zhzhang/PG/RBP/SPhsmm_3gene_RBPrichness.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#plot(排除基因间区，以及未被检测到与RBP结合的基因)
phs <- ggplot(data=filter(filter(ALLSP_RBP_NUM,type!="Random intergenic" & RBPnum!=0),sp=="Human"), aes(x=type,y=log10(RBPnum)))+
  geom_boxplot(fatten = 3,outlier.alpha = 0,width=0.5,notch=T,aes(fill=type))+
  geom_signif(map_signif_level=T,y_position=c(2.5,2.25),
              comparisons = list(c("Protein-coding","Pseudogene-derived lncRNA"),
                                 c("Non-pseudogene-derived lncRNA","Pseudogene-derived lncRNA")
              ))+
  scale_fill_manual(values=c("#BB5A5D","#7197AD","#FAA465","#8491B4"),
                    limits=c("Protein-coding","Pseudogene-derived lncRNA",
                             "Non-pseudogene-derived lncRNA"))+
  theme_half_open()+
  scale_y_continuous(breaks = c(0,1,2),labels = c("0","1","2"))+
  scale_x_discrete(labels = c("Protein-coding","Pseudogene-\nderived lncRNA",
                              "Non-pseudogene-\nderived lncRNA"))+
  labs(y = expression("l"*"o"*"g"[10]*"("*"R"*"B"*"P"~"r"*"i"*"c"*"h"*"n"*"e"*"s"*"s"*")"),
       x =NULL,fill = NULL,color = NULL)+
  theme(axis.title = element_text(size = 20),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),legend.position = "none") +
  theme(axis.text.x = element_text(angle =40)) +
  theme(axis.text.x = element_text(hjust=1,vjust = 1))
ggsave("/home/zhzhang/PG/RBP/Homo_sapiens.3gene_RBPrichness.png",
       phs,width = 4, height = 5,dpi=1200, units = "in", device='png',bg = "transparent")





```


##### 根据基因具有的RBP种类及其结合长度信息文件，输出基因RBP结合长度矩阵（每行代表基因，每列代表一类RBP）
```r
#函数根据基因&基因间区具有的RBP种类及其结合长度信息文件，输出基因&基因间区RBP结合长度矩阵（每行代表基因，每列代表一类RBP）
#a输入基因具有的RBP种类及其结合长度信息文件，b输出基因RBP结合长度矩阵文件（每行代表基因，每列代表一类RBP）
getgene_RBP_matrix <- function(a,b){
  #导入基因具有的RBP种类及其结合长度信息文件
  pcgApg_RBP <- read.delim(a)
  #统计全部RBP类型
  RBP_class <- select(pcgApg_RBP,RBP)%>%
    distinct(RBP)
  #统计全部geneid
  geneid <- select(pcgApg_RBP,geneid)%>%
    distinct(geneid)
  #先循环获得每行为一类RBP，每列代表基因的矩阵
  gene_RBP_matrix <- RBP_class
  genesample_num <- nrow(geneid)
  for (i in c(1:nrow(geneid))) {
    #提取出第i个基因的RBP结合长度信息
    genei <- geneid[i,1]#i
    genei_RBP <- filter(pcgApg_RBP,geneid==genei)%>%
      select(-1)
    colnames(genei_RBP)[2] <- genei
    #将基因i的各RBP结合长度信息以列增添的形式循环添加到总矩阵中
    gene_RBP_matrix <- left_join(gene_RBP_matrix,genei_RBP,by="RBP")
    gene_RBP_matrix[is.na(gene_RBP_matrix[,i+1])==T,i+1] <- 0#i+1
    #输出进度
    print(i/genesample_num)
  }
  #转置生成基因RBP结合长度矩阵（每行代表基因，每列代表一类RBP）
  gene_RBP_matrix_t <- column_to_rownames(gene_RBP_matrix,"RBP")%>%
    t()%>%
    data.frame()
  #储存
  data.table::fwrite(gene_RBP_matrix_t,file =b,sep = '\t',row.names = T,quote = F,col.names = T)
}

#小鼠
getgene_RBP_matrix(a="/home/zhzhang/PG/RBP/Mus_musculus.allgeneexonAintergenic.RBP.txt",
                   b="/home/zhzhang/PG/RBP/Mus_musculus.allgeneexonAintergenic.RBP.matrix.txt")
#人
getgene_RBP_matrix(a="/home/zhzhang/PG/RBP/Homo_sapiens.allgeneexonAintergenic.RBP.txt",
                   b="/home/zhzhang/PG/RBP/Homo_sapiens.allgeneexonAintergenic.RBP.matrix.txt")



```


##### 对比蛋白编码/lncRNA基因转录本以及基因间区的RBP多样性diversity
```r
#函数根据RBP结合比例矩阵文件，统计RBP多样性diversity
#a输入基因&基因间区RBP结合比例矩阵文件，b输入物种名，c输入基因分类文件,d输入基因间区文件
getRBP_diversity <- function(a,b,c,d){
  #导入RBP结合比例矩阵
  gene_RBP_matrix <- read.delim(a, row.names=1)
  #计算每个基因的RBP多样性
  gene_RBP_diversity <- data.frame("geneid"=rownames(gene_RBP_matrix),
                                    "simpson"=vegan::diversity(gene_RBP_matrix, index = "simpson"),
                                    "shannon"=vegan::diversity(gene_RBP_matrix, index = "shannon"))
  #导入基因分类文件
  geneid_class <- read.delim(c)
  #导入基因间区分类
  intergenic <- read.delim(d, header=FALSE)%>%
    select(1)%>%
    mutate(type="Random intergenic")
  colnames(intergenic)[1] <- "geneid"
  #合并分类信息
  allclass <- rbind(geneid_class,intergenic)
  #基因分类以及RBP多样性信息合并
  geneid_class_RBP <- left_join(allclass,gene_RBP_diversity,by="geneid")%>%
    filter(type!="Interference lncRNA")%>%
    mutate(sp=b)
  geneid_class_RBP$simpson[is.na(geneid_class_RBP$simpson)==T] <- 0
  geneid_class_RBP$shannon[is.na(geneid_class_RBP$shannon)==T] <- 0
  return(geneid_class_RBP)
}


#人
hs_RBP_diversity <- getRBP_diversity(a="/home/zhzhang/PG/RBP/Homo_sapiens.allgeneexonAintergenic.RBP.matrix.txt",
                                       b="Human",
                                       c="~/PG/RNAseq/Homo_sapiens/Homo_sapiens.geneid_class.txt",
                                     d="~/PG/Evolution/conserve/Homo_sapiens.20000random_3kb_intergenic.phastCons.txt")

#合并所有物种结果
ALLSP_RBP_diversity <- hs_RBP_diversity
ALLSP_RBP_diversity$type <- factor(ALLSP_RBP_diversity$type,levels = c("Protein-coding","Pseudogene-derived lncRNA",
                                                                       "Non-pseudogene-derived lncRNA","Random intergenic"))
data.table::fwrite(ALLSP_RBP_diversity,file ="/home/zhzhang/PG/RBP/SPhsmm_3gene_RBPdiversity.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#统计(排除基因间区，以及未被检测到与RBP结合的基因)
tj <- group_by(filter(ALLSP_RBP_diversity,type!="Random intergenic" & simpson!=0 & shannon!=0),sp,type)%>%
  summarise(simpsonmean=mean(simpson),simpsonmedian=median(simpson),
            shannonmean=mean(shannon),shannonmedian=median(shannon))
data.table::fwrite(tj,file ="/home/zhzhang/PG/RBP/SPhsmm_3gene_RBPdiversity.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#plot(排除基因间区，以及未被检测到与RBP结合的基因)
phsshannon <- ggplot(data=filter(filter(ALLSP_RBP_diversity,type!="Random intergenic" & simpson!=0 & shannon!=0),sp=="Human"), aes(x=type,y=shannon))+
  geom_boxplot(fatten = 3,outlier.alpha = 0,width=0.5,notch=T,aes(fill=type))+
  geom_signif(map_signif_level=T,y_position=c(4.9,4.5),
              comparisons = list(c("Protein-coding","Pseudogene-derived lncRNA"),
                                 c("Non-pseudogene-derived lncRNA","Pseudogene-derived lncRNA")
              ))+
  scale_fill_manual(values=c("#BB5A5D","#7197AD","#FAA465","#8491B4"),
                    limits=c("Protein-coding","Pseudogene-derived lncRNA",
                             "Non-pseudogene-derived lncRNA"))+
  theme_half_open()+
  scale_y_continuous(breaks = c(0,1,2,3,4,5))+
  scale_x_discrete(labels = c("Protein-coding","Pseudogene-\nderived lncRNA",
                              "Non-pseudogene-\nderived lncRNA"))+
  labs(y = "Shannon-Wiener index",
       x =NULL,fill = NULL,color = NULL)+
  theme(axis.title = element_text(size = 20),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),legend.position = "none") +
  theme(axis.text.x = element_text(angle =40)) +
  theme(axis.text.x = element_text(hjust=1,vjust = 1))
ggsave("/home/zhzhang/PG/RBP/Homo_sapiens.3gene_RBPdiversity.shannon.png",
       phsshannon,width = 3.5, height = 5,dpi=1200, units = "in", device='png',bg = "transparent")

phssimpson <- ggplot(data=filter(filter(ALLSP_RBP_diversity,type!="Random intergenic" & simpson!=0 & shannon!=0),sp=="Human"), aes(x=type,y=simpson))+
  geom_boxplot(fatten = 3,outlier.alpha = 0,width=0.5,notch=T,aes(fill=type))+
  geom_signif(map_signif_level=T,y_position=c(1.1,1),
              comparisons = list(c("Protein-coding","Pseudogene-derived lncRNA"),
                                 c("Non-pseudogene-derived lncRNA","Pseudogene-derived lncRNA")
              ))+
  scale_fill_manual(values=c("#BB5A5D","#7197AD","#FAA465","#8491B4"),
                    limits=c("Protein-coding","Pseudogene-derived lncRNA",
                             "Non-pseudogene-derived lncRNA"))+
  theme_half_open()+
  scale_y_continuous(breaks = c(0,0.25,0.5,0.75,1))+
  scale_x_discrete(labels = c("Protein-coding","Pseudogene-\nderived lncRNA",
                              "Non-pseudogene-\nderived lncRNA"))+
  labs(y = "(1 - Simpson index)",
       x =NULL,fill = NULL,color = NULL)+
  theme(axis.title = element_text(size = 20),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),legend.position = "none") +
  theme(axis.text.x = element_text(angle =40)) +
  theme(axis.text.x = element_text(hjust=1,vjust = 1))
ggsave("/home/zhzhang/PG/RBP/Homo_sapiens.3gene_RBPdiversity.simpson.png",
       phssimpson,width = 3.5, height = 5,dpi=1200, units = "in", device='png',bg = "transparent")


```
##### pgdlnc基因与母基因之间的RBP相似性
```r
#R函数给指定物种每个pglnc基因分配随机的蛋白编码基因
#最终生成至少有一种RBP结合的pglnc-母基因，pglnc-随机蛋白编码基因，随机lnc-母基因的基因id对应文件
#（a输入物种每个基因id的RBP丰富度文件,b输入pglnc和其对应的母基因文件，c输出全部需要计算表达相关性的三类基因对）
pgrandompcg <- function(a,b,c){
  #导入每个基因id的RBP丰富度文件，筛选出与至少一种RBP结合的基因
  geneid_class <- read.delim(a)%>%
    filter(sp=="Human" & RBPnum>0)%>%
    select(1,2)
  #获取pglnc基因id
  pglncid <- filter(geneid_class,type=="Pseudogene-derived lncRNA")%>%
    select(1)
  #获取蛋白编码基因id
  codingid <- filter(geneid_class,type=="Protein-coding")%>%
    select(1)
  #获取npglnc基因id
  npglncid <- filter(geneid_class,type=="Non-pseudogene-derived lncRNA")%>%
    select(1)
  #导入pglnc和其对应的母基因文件,仅保留pglnc基因和母基因都与至少一种RBP结合的基因对
  pgdlncRNA_transtarget <- read.delim(b)
  colnames(pgdlncRNA_transtarget)[1] <- c("geneid")
  pgdlncRNA_transtarget <- inner_join(pgdlncRNA_transtarget,pglncid,by="geneid")
  colnames(pgdlncRNA_transtarget) <- c("pglncid","geneid")
  pgdlncRNA_transtarget <- inner_join(pgdlncRNA_transtarget,codingid,by="geneid")
  colnames(pgdlncRNA_transtarget) <- c("geneid","target")
  pgdlncRNA_transtarget <- mutate(pgdlncRNA_transtarget,type="Pseudogene-derived lncRNA vs Parent gene of pseudogene")
  #获取需要分配随机蛋白基因的pglnc基因数量
  pglncnum <- nrow(distinct(pgdlncRNA_transtarget,geneid))
  #生成pglnc-随机蛋白编码基因对数量>3000
  pglnc_suiji_pcg <- data.frame()
  while (nrow(pglnc_suiji_pcg)<3000) {
    #随机抽取5倍(pglnc基因数量)个蛋白编码基因ID
    chou_pcg <- sample_n(data.frame(target=setdiff(codingid$geneid,distinct(pgdlncRNA_transtarget,target)$target)),pglncnum*5)
    #给pglnc分配随机蛋白基因
    pglnc_suiji_pcg_while <- cbind(rbind(distinct(pgdlncRNA_transtarget,geneid),
                                         distinct(pgdlncRNA_transtarget,geneid),
                                         distinct(pgdlncRNA_transtarget,geneid),
                                         distinct(pgdlncRNA_transtarget,geneid),
                                         distinct(pgdlncRNA_transtarget,geneid)),chou_pcg)%>%
      mutate(type="Pseudogene-derived lncRNA vs Random protein-coding gene")
    pglnc_suiji_pcg <- rbind(pglnc_suiji_pcg,pglnc_suiji_pcg_while)
  }
  #获取需要分配随机lnc基因的母基因数量
  parentnum <- nrow(distinct(pgdlncRNA_transtarget,target))
  #生成随机lnc-上述母基因对数量>3000
  suiji_lnc_parent <- data.frame()
  while (nrow(suiji_lnc_parent)<3000) {
    #随机抽取5倍(母基因数量)个lnc基因ID
    chou_lncg <- sample_n(npglncid,parentnum*5)
    #给母基因分配随机lnc基因
    suiji_lnc_parent_while <- cbind(chou_lncg,rbind(distinct(pgdlncRNA_transtarget,target),
                                                    distinct(pgdlncRNA_transtarget,target),
                                                    distinct(pgdlncRNA_transtarget,target),
                                                    distinct(pgdlncRNA_transtarget,target),
                                                    distinct(pgdlncRNA_transtarget,target)))%>%
      mutate(type="Random lncRNA vs Parent gene of pseudogene")
    suiji_lnc_parent <- rbind(suiji_lnc_parent,suiji_lnc_parent_while)
  }
  #合并全部需要计算表达相关性的基因对
  allgenepair <- rbind(pgdlncRNA_transtarget,pglnc_suiji_pcg,suiji_lnc_parent)
  #储存
  data.table::fwrite(allgenepair,file =c,sep = '\t',row.names = F,quote = F,col.names = T)
}
#人类
pgrandompcg(a = "/home/zhzhang/PG/RBP/SPhsmm_3gene_RBPrichness.txt",
            b = "~/PG/lncRNA_class/Homo_sapiens.pgdlncRNA_transtarget.txt",
            c = "/home/zhzhang/PG/RBP/Homo_sapiens.three_genepair_needRBPdistance.txt")



#函数根据RBP结合比例矩阵文件，计算三类基因对之间的RBP Bray-Curtis similarity index
#(a输入RBP结合比例矩阵文件，b输入基因对文件，c输入物种名
getpgpcgRBPdistance <- function(a,b,c){
  #导入RBP结合比例矩阵文件
  gene_RBP_matrix <- read.delim(a)
  #导入基因对
  three_genepair_needdis <- read.delim(b)
  #循环计算每一对基因对的RBP相似性
  genepair_disdata <- data.frame()
  for (i in c(1:nrow(three_genepair_needdis))) {
    #提取两个基因ID
    geneid <- three_genepair_needdis[i,1]#
    targetid <- three_genepair_needdis[i,2]#
    #提取基因对RBP矩阵
    genepair_RBP <- filter(gene_RBP_matrix,X==geneid | X==targetid)%>%
      column_to_rownames("X")
    #计算基因对的RBP相似性
    if (nrow(genepair_RBP)==2) {
      vegdist <- vegan::vegdist(genepair_RBP, binary=F,method="bray")
      #每对基因对的RBP相似性存入新的dataframe
      genepair_disdata[i,1] <- geneid
      genepair_disdata[i,2] <- targetid
      genepair_disdata[i,3] <- three_genepair_needdis[i,3]
      genepair_disdata[i,4] <- 1-vegdist
    }
    #输出完成量
    print(i/nrow(three_genepair_needdis)*100)
  }
  #基因对的RBP相似性dataframe清除无法计算的对(dis=NA)
  genepair_disdata <- genepair_disdata[is.na(genepair_disdata[,4])==F,]
  colnames(genepair_disdata) <- c("geneid","target","type","similarity")
  #增加物种信息
  genepair_disdata <- mutate(genepair_disdata,sp=c)
}
#人
hs_pgpcgRBPsimi <- getpgpcgRBPdistance(a="/home/zhzhang/PG/RBP/Homo_sapiens.allgeneexonAintergenic.RBP.matrix.txt",
                                       b="/home/zhzhang/PG/RBP/Homo_sapiens.three_genepair_needRBPdistance.txt",
                                       c="Human")
#物种RBP相似性数据储存
hs_pgpcgRBPsimi$type <- factor(hs_pgpcgRBPsimi$type,
                                 levels = c("Pseudogene-derived lncRNA vs Parent gene of pseudogene",
                                            "Pseudogene-derived lncRNA vs Random protein-coding gene",
                                            "Random lncRNA vs Parent gene of pseudogene"))
data.table::fwrite(hs_pgpcgRBPsimi,file ="/home/zhzhang/PG/RBP/Homo_sapiens.three_genepair_RBPsimidata.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#统计储存
tj <- group_by(hs_pgpcgRBPsimi,sp,type)%>%
  summarise(mean=mean(similarity),median=median(similarity),sd=sd(similarity))
data.table::fwrite(tj,file ="/home/zhzhang/PG/RBP/Homo_sapiens.three_genepair_RBPsimidata.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#人
phs <- ggplot(data = filter(hs_pgpcgRBPsimi,sp=="Human"),aes(x=type,y=similarity))+
  geom_boxplot(notch =T,fatten = 3,width=0.5,outlier.alpha = 0,aes(fill=type))+
  geom_signif(map_signif_level=T,y_position = c(0.56,0.62),
              comparisons=list(c("Pseudogene-derived lncRNA vs Parent gene of pseudogene","Pseudogene-derived lncRNA vs Random protein-coding gene"),
                               c("Pseudogene-derived lncRNA vs Parent gene of pseudogene","Random lncRNA vs Parent gene of pseudogene")
              ))+
  scale_fill_manual(values = c("#B20339","#4E6AB5","#614499"))+
  theme_half_open()+
  coord_cartesian(ylim = c(0, 0.7))+
  scale_y_continuous(breaks = c(0,0.2,0.4,0.6),labels = c("0","0.2","0.4","0.6"))+
  scale_x_discrete(labels = c("","",""))+
  labs(x = NULL, y ="Bray-Curtis\nsimilarity index of RBP",fill = NULL)+
  theme(axis.title = element_text(size = 20),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 13)) +
  theme(axis.ticks.x = element_line(size = 0)) +
  theme(legend.text = element_text(size = 14))
ggsave("/home/zhzhang/PG/RBP/Homo_sapiens.pgdlnc_parent_RBPsimi.png", 
       phs,width = 9.5, height = 5,dpi=1200, units = "in", device='png',bg = "transparent")



```
##### pgdlnc基因与母基因之间共享的RBP数量及比例
```r
#函数根据RBP结合比例矩阵文件，计算三类基因对之间 共享的RBP种类数
#(a输入RBP结合比例矩阵文件，b输入基因对文件，c输入物种名
getpgpcgRBPdistance <- function(a,b,c){
  #导入RBP结合比例矩阵文件
  gene_RBP_matrix <- read.delim(a)
  #导入基因对
  three_genepair_needdis <- read.delim(b)
  #循环计算每一对基因对的RBP相似性
  genepair_disdata <- data.frame()
  for (i in c(1:nrow(three_genepair_needdis))) {
    #提取两个基因ID
    geneid <- three_genepair_needdis[i,1]#
    targetid <- three_genepair_needdis[i,2]#
    #提取基因对RBP矩阵
    genepair_RBP <- filter(gene_RBP_matrix,X==geneid | X==targetid)%>%
      column_to_rownames("X")
    #转置并筛选出两基因shared RBP
    genepair_RBP <- data.frame(t(genepair_RBP))%>%
      filter(get(geneid)>0 & get(targetid)>0)
    #计算基因对共享的RBP数目
    if (ncol(genepair_RBP)==2) {
      vegdist <- nrow(genepair_RBP)
      #每对基因对的RBP相似性存入新的dataframe
      genepair_disdata[i,1] <- geneid
      genepair_disdata[i,2] <- targetid
      genepair_disdata[i,3] <- three_genepair_needdis[i,3]
      genepair_disdata[i,4] <- vegdist
    }
    #输出完成量
    print(i/nrow(three_genepair_needdis)*100)
  }
  #基因对的RBP相似性dataframe清除无法计算的对(dis=NA)
  genepair_disdata <- genepair_disdata[is.na(genepair_disdata[,4])==F,]
  colnames(genepair_disdata) <- c("geneid","target","type","similarity")
  #增加物种信息
  genepair_disdata <- mutate(genepair_disdata,sp=c)
}
#人
hs_pgpcgRBPsimi <- getpgpcgRBPdistance(a="/home/zhzhang/PG/RBP/Homo_sapiens.allgeneexonAintergenic.RBP.matrix.txt",
                                       b="/home/zhzhang/PG/RBP/Homo_sapiens.three_genepair_needRBPdistance.txt",
                                       c="Human")
#物种RBP相似性数据储存
hs_pgpcgRBPsimi$type <- factor(hs_pgpcgRBPsimi$type,
                               levels = c("Pseudogene-derived lncRNA vs Parent gene of pseudogene",
                                          "Pseudogene-derived lncRNA vs Random protein-coding gene",
                                          "Random lncRNA vs Parent gene of pseudogene"))
data.table::fwrite(hs_pgpcgRBPsimi,file ="/home/zhzhang/PG/RBP/Homo_sapiens.three_genepair_sharedRBPnum.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#统计储存
tj <- group_by(hs_pgpcgRBPsimi,sp,type)%>%
  summarise(mean=mean(similarity),median=median(similarity),sd=sd(similarity))
data.table::fwrite(tj,file ="/home/zhzhang/PG/RBP/Homo_sapiens.three_genepair_sharedRBPnum.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#人
phs <- ggplot(data = filter(hs_pgpcgRBPsimi,sp=="Human"),aes(x=type,y=similarity))+
  geom_boxplot(notch =T,fatten = 3,width=0.5,outlier.alpha = 0,aes(fill=type))+
  geom_signif(map_signif_level=T,y_position = c(30,34),tip_length = 0.01,
              comparisons=list(c("Pseudogene-derived lncRNA vs Parent gene of pseudogene","Pseudogene-derived lncRNA vs Random protein-coding gene"),
                               c("Pseudogene-derived lncRNA vs Parent gene of pseudogene","Random lncRNA vs Parent gene of pseudogene")
              ))+
  scale_fill_manual(values = c("#B20339","#4E6AB5","#614499"))+
  theme_half_open()+
  coord_cartesian(ylim = c(0, 40))+
  scale_x_discrete(labels = c("","",""))+
  labs(x = NULL, y ="Number of shared RBPs",fill = NULL)+
  theme(axis.title = element_text(size = 20),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 13)) +
  theme(axis.ticks.x = element_line(size = 0)) +
  theme(legend.text = element_text(size = 14))
ggsave("/home/zhzhang/PG/RBP/Homo_sapiens.pgdlnc_parent_sharedRBPnum.png", 
       phs,width = 9.5, height = 5,dpi=1200, units = "in", device='png',bg = "transparent")


#统计至少共享i个RBP的基因对占比
tjratio_he <- data.frame()
for (i in c(1:10)) {
  cutoff <- i
  forratio <- mutate(hs_pgpcgRBPsimi,share=0)
  forratio$share[forratio$similarity>=cutoff] <- 1
  tjratio <- group_by(forratio,type)%>%
    summarise(num=n(),sharenum=sum(share))%>%
    mutate(nonum=num-sharenum,percent=sharenum*100/num)%>%
    mutate(cutoffvalue=cutoff)
  tjratio_he <- rbind(tjratio_he,tjratio)
}
#储存
data.table::fwrite(tjratio_he,file ="/home/zhzhang/PG/RBP/Homo_sapiens.pgdlnc_parent_sharedRBPnum_percent.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#plot
ph <- ggplot(data = tjratio_he,aes(x=cutoffvalue,y=percent))+
  geom_point(aes(color=type),size=2)+
  geom_line(aes(group=type,color=type))+
  scale_color_manual(values = c("#B20339","#4E6AB5","#614499"))+
  theme_half_open()+
  scale_x_continuous(breaks = c(2,4,6,8,10),labels = paste("≥",c(2,4,6,8,10),sep=""))+
  scale_y_continuous(breaks = c(15,30,45,60,75,90))+
  coord_cartesian(ylim = c(15, 90))+
  labs(x = "Number of shared RBPs", y ="Percentage (%)",colour = NULL)+
  theme(axis.title = element_text(size = 20),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 15),
        legend.position = "none", legend.direction = "vertical")
ggsave("/home/zhzhang/PG/RBP/Homo_sapiens.pgdlnc_parent_sharedRBPnum_percent.png", 
       ph,width = 5, height = 5,dpi=1200, units = "in", device='png',bg = "transparent")



```











##### 通过共表达基因推断pglnc功能
```r
#分裂lncid（1000 line per file）
cat /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/Homo_sapiens.spexp_pgdlncgene.id.txt /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/Homo_sapiens.spexp_npgdlncgene.id.txt |split -l 1000 -d - /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/lncid/lncid
ls /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/lncid/ > /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/split.lncid.txt


#分裂pcgid（10 line per file）
split -l 10 -d /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/Homo_sapiens.spexp_pcgene.id.txt /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/pcgid/pcgid
ls /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/pcgid/ > /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/split.pcgid.txt


#找到全部共表达的lnc-pcg基因对
#R
library(magrittr)
#生成参数列表
option_list <- list(optparse::make_option(
  opt_str ="--lncid",
  type = "character",
  default = NULL,
  help="splited spexp protein-coding geneid txt file"),
  optparse::make_option(
    opt_str ="--pcgid",
    type = "character",
    default = NULL,
    help="splited spexp protein-coding geneid txt file"),
  optparse::make_option(
    opt_str ="--outputone",
    type = "character",
    default = NULL,
    help="splited out"
  ),
  optparse::make_option(
    opt_str ="--outputtwo",
    type = "character",
    default = NULL,
    help="splited and filtered out"
  ))
#将参数列表传递给一个变量
args <- optparse::parse_args(optparse::OptionParser(option_list=option_list))
getgenepaircor <- function(a,b,c,d,e){
  #导入表达的lnc基因list（1000）
  lncid <- read.delim(b)
  colnames(lncid) <- "geneid"
  #导入表达的蛋白编码基因list（10）
  pcgid <- read.delim(c)
  colnames(pcgid) <- "geneid"
  #导入基因表达量TPM矩阵
  allsample_TPM <- data.table::fread(a)%>%
    data.frame()
  colnames(allsample_TPM)[1] <- "geneid"
  #lnc基因表达矩阵
  lncid_allsample_TPM <- dplyr::inner_join(allsample_TPM,lncid,by="geneid")%>%
    tibble::column_to_rownames("geneid")%>%
    t()%>%
    data.frame()
  #蛋白基因表达矩阵
  pcgid_allsample_TPM <- dplyr::inner_join(allsample_TPM,pcgid,by="geneid")%>%
    tibble::column_to_rownames("geneid")%>%
    t()%>%
    data.frame()
      #计算基因对的表达相关性
      cor <- psych::corr.test(x=lncid_allsample_TPM,
                              y=pcgid_allsample_TPM,
                              method="pearson",adjust="holm")
      perrson <- cor[["r"]]%>%
        data.frame()%>%
        tibble::rownames_to_column("geneid")%>%
        tidyr::gather("pcgid","cor",2:(nrow(pcgid)+1))
      pvalue <- cor[["p.adj"]]%>%
        data.frame()%>%
        tibble::rownames_to_column("geneid")%>%
        tidyr::gather("pcgid","pvalue",2:(nrow(pcgid)+1))%>%
        dplyr::select(pvalue)
      corandpadj <- cbind(perrson,pvalue)
  #保存
  data.table::fwrite(corandpadj,file =d,
                     sep = '\t',row.names = F,quote = F,col.names = F)
  #筛选
  filtercorlong <- dplyr::filter(corandpadj,pvalue<0.01 & abs(cor)>0.75)
  data.table::fwrite(filtercorlong,file =e,
                     sep = '\t',row.names = F,quote = F,col.names = F)
}
getgenepaircor(a="/home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/cbindsample_TPM.txt",
               b=args$lncid,
               c=args$pcgid,
               d=args$outputone,
               e=args$outputtwo)

#parallel 并行执行R脚本
#人
cat "/home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/split.lncid.txt"|while read i
do
echo "${i}"
cat "/home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/split.pcgid.txt" | parallel -j 120 --verbose Rscript "/home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/sh/hs.R" --lncid /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/lncid/${i} --pcgid /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/pcgid/{} --outputone /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/out1/${i}.{}.out --outputtwo /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/out2/${i}.{}.out
done

#合并cor并行结果
#人
cat /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/out2/* > /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/merge/Homo_sapiens.spexp.lnc_pcgene.cordata.filter.txt
#获取全部lnc的coexp蛋白编码基因id（作为证明pglnc通路更focus在精子发生时，富集的bgd）
awk '{print $2}' "/home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/merge/Homo_sapiens.spexp.lnc_pcgene.cordata.filter.txt"|sort|uniq > /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/merge/Homo_sapiens.spexp.lnc_coexp_pcg.bgd.id.txt
#提取pglnc-pcg共表达对
grep -w -Ff "/home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/Homo_sapiens.spexp_pgdlncgene.id.txt" "/home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/merge/Homo_sapiens.spexp.lnc_pcgene.cordata.filter.txt" > "/home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/merge/Homo_sapiens.spexp.pglnc_pcgene.cordata.filter.txt"
#all pglnc coexp pcgid
awk '{print $2}' /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/merge/Homo_sapiens.spexp.pglnc_pcgene.cordata.filter.txt|sort|uniq > /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/merge/Homo_sapiens.spexp.pglnc_coexp.id.txt 
#提取npglnc-pcg共表达对
grep -v -w -Ff "/home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/Homo_sapiens.spexp_pgdlncgene.id.txt" "/home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/merge/Homo_sapiens.spexp.lnc_pcgene.cordata.filter.txt" > "/home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/merge/Homo_sapiens.spexp.npglnc_pcgene.cordata.filter.txt"
#all npglnc coexp pcgid
awk '{print $2}' /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/merge/Homo_sapiens.spexp.npglnc_pcgene.cordata.filter.txt|sort|uniq > /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/merge/Homo_sapiens.spexp.npglnc_coexp.id.txt 




###证明pglnc通路更focus在精子发生
#导入两类lnc的靶基因
pgcoexp <- read.table("/home/zhzhang/PG/RNAseq/Homo_sapiens/addtarget_new/Homo_sapiens.spexp.pglnc_coexp.id.txt", quote="\"", comment.char="")
colnames(pgcoexp) <- "geneid"
npgcoexp <- read.table("/home/zhzhang/PG/RNAseq/Homo_sapiens/addtarget_new/Homo_sapiens.spexp.npglnc_coexp.id.txt", quote="\"", comment.char="")
colnames(npgcoexp) <- "geneid"
#导入背景基因集
bgd <- read.table("/home/zhzhang/PG/RNAseq/Homo_sapiens/addtarget_new/Homo_sapiens.spexp.lnc_coexp_pcg.bgd.id.txt",
                  quote="\"", comment.char="")
colnames(bgd) <- "geneid"
#pglnc all靶基因富集信息
pgcoexp_GO <- clusterProfiler::enrichGO(gene = pgcoexp$geneid, universe=bgd$geneid,OrgDb = 'org.Hs.eg.db', keyType = "ENSEMBL", ont = 'BP', pAdjustMethod = 'fdr', pvalueCutoff = 0.05, qvalueCutoff = 1)
pgcoexp_GO_result <- pgcoexp_GO@result
pgcoexp_GO_result <- pgcoexp_GO_result[which(pgcoexp_GO_result$Description=="germ cell development"|
                                               pgcoexp_GO_result$Description=="meiotic cell cycle"|
                                               pgcoexp_GO_result$Description=="meiotic cell cycle process"|
                                               pgcoexp_GO_result$Description=="meiotic nuclear division"|
                                               pgcoexp_GO_result$Description=="cilium organization"|
                                               pgcoexp_GO_result$Description=="cilium assembly"|
                                               pgcoexp_GO_result$Description=="cilium movement"|
                                               pgcoexp_GO_result$Description=="cilium-dependent cell motility"|
                                               pgcoexp_GO_result$Description=="spermatid differentiation"|
                                               pgcoexp_GO_result$Description=="spermatid development"),]

pgcoexp_GO_result <- separate(pgcoexp_GO_result,GeneRatio,c("no","all"),sep="/",remove = F)%>%
  mutate(all=as.numeric(all))%>%
  mutate(Generatio=Count/all)%>%
  select(-no,-all)%>%
  separate(BgRatio,c("BgRatio1","BgRatio2"),sep="/",remove = F)%>%
  mutate(BgRatio1=as.numeric(BgRatio1))%>%
  mutate(BgRatio2=as.numeric(BgRatio2))%>%
  mutate(Bgratio=BgRatio1/BgRatio2)%>%
  select(-BgRatio1,-BgRatio2)%>%
  mutate(Description=Hmisc::capitalize(Description))%>%
  mutate(ES=Generatio/Bgratio)%>%
  mutate(type="Pseudogene-derived lncRNA")
#npglnc all靶基因富集信息
npg_GO <- clusterProfiler::enrichGO(gene = npgcoexp$geneid, universe=bgd$geneid,OrgDb = 'org.Hs.eg.db', keyType = "ENSEMBL", ont = 'BP', pAdjustMethod = 'fdr', pvalueCutoff = 0.05, qvalueCutoff = 1)
npg_GO_result <- npg_GO@result
npg_GO_result <- npg_GO_result[which(npg_GO_result$Description=="germ cell development"|
                           npg_GO_result$Description=="meiotic cell cycle"|
                           npg_GO_result$Description=="meiotic cell cycle process"|
                           npg_GO_result$Description=="meiotic nuclear division"|
                           npg_GO_result$Description=="cilium organization"|
                           npg_GO_result$Description=="cilium assembly"|
                           npg_GO_result$Description=="cilium movement"|
                           npg_GO_result$Description=="cilium-dependent cell motility"|
                           npg_GO_result$Description=="spermatid differentiation"|
                           npg_GO_result$Description=="spermatid development"),]
npg_GO_result <- separate(npg_GO_result,GeneRatio,c("no","all"),sep="/",remove = F)%>%
  mutate(all=as.numeric(all))%>%
  mutate(Generatio=Count/all)%>%
  select(-no,-all)%>%
  separate(BgRatio,c("BgRatio1","BgRatio2"),sep="/",remove = F)%>%
  mutate(BgRatio1=as.numeric(BgRatio1))%>%
  mutate(BgRatio2=as.numeric(BgRatio2))%>%
  mutate(Bgratio=BgRatio1/BgRatio2)%>%
  select(-BgRatio1,-BgRatio2)%>%
  mutate(Description=Hmisc::capitalize(Description))%>%
  mutate(ES=Generatio/Bgratio)%>%
  mutate(type="Non-pseudogene-derived lncRNA")
#hebing
lnche <- rbind(pgcoexp_GO_result,npg_GO_result)
lnche$type <- factor(lnche$type,levels=c("Non-pseudogene-derived lncRNA",
                                         "Pseudogene-derived lncRNA"))
lnche$Description <- factor(lnche$Description,levels=c("Spermatid development",
                                                       "Spermatid differentiation",
                                                       "Cilium-dependent cell motility",
                                                       "Cilium movement",
                                                       "Cilium assembly",
                                                       "Cilium organization",
                                                       "Meiotic nuclear division",
                                                       "Meiotic cell cycle process",
                                                       "Meiotic cell cycle",
                                                       "Germ cell development"))
#储存
data.table::fwrite(lnche,file ="/home/zhzhang/PG/RNAseq/Homo_sapiens/addtarget_new/Homo_sapiens.spexp.allpglnc_npglnc_coexpgcgene.bgdenrich.txt",
                   sep = '\t',row.names = F,quote = F,col.names = T)
#富集dotplot
lnche <- mutate(lnche,anno="N.S.")
lnche$anno[lnche$p.adjust < 0.05] <- "*"
lnche$anno[lnche$p.adjust < 0.01] <- "**"
lnche$anno[lnche$p.adjust < 0.001] <- "***"
pp1 <- ggplot(data = lnche,aes(x=type,y=Description))+
  geom_point(aes(fill=-log10(p.adjust),size=ES),shape=21,color="black",
             position =position_nudge(x=0.15))+
  geom_text(aes(label=anno),nudge_x =-0.15 )+
  geom_hline(yintercept=c(1.5:9.5),color="gray")+
  geom_vline(xintercept=c(1.5),color="gray")+
  scale_fill_gradient(low="white",high="#C12039")+
  scale_x_discrete(labels=c("Non-pseudogene-\nderived lncRNA",
                            "Pseudogene-derived\nlncRNA"))+
  theme_half_open()+
  scale_y_discrete(position = "right")+
  labs(x = NULL, y =NULL,size="Enrichment score")+
  theme(axis.text.x = element_text(angle =45))+
  theme(axis.text.x = element_text(vjust = 1,hjust=1))+
  theme(axis.title = element_text(size = 15),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),legend.text = element_text(size = 14)) +
  theme(panel.background = element_rect(colour = "black",size=1)) +
  theme(legend.position = "left")
ggsave("/home/zhzhang/PG/RNAseq/plot/Homo_sapiens.spexp.allpglnc_npglnc_coexpgcgene.bgdenrich.pdf", 
       pp1,width = 7, height = 6,dpi=1200, units = "in", device='pdf',bg = "transparent")












#TOP20共表达基因用于构建PPI网络
#提取每个pglnc top20共表达pcg
a <- "/home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/merge/Homo_sapiens.spexp.pglnc_pcgene.cordata.filter.txt"
c <- "/home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/merge/Homo_sapiens.spexp.pglnc_pcgene.coexppair.top20.txt"
#导入基因对cor
pglnc_pcgene_cordata <- read.delim(a, header=FALSE)
pglnc_pcgene_cordata_g20 <- group_by(pglnc_pcgene_cordata,V1)%>%
  top_n(20,V3)
#储存筛选后的pglnc及其靶基因
colnames(pglnc_pcgene_cordata_g20) <- c("lncgid","targetid","cor","padj")
data.table::fwrite(pglnc_pcgene_cordata_g20,
                   file =c,sep = '\t',row.names = F,quote = F,col.names = T)
#pglnc top20共表达pcgid
tail -n +2 /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/merge/Homo_sapiens.spexp.pglnc_pcgene.coexppair.top20.txt|awk '{print $2}'|sort|uniq > /home/zhzhang/PG/RNAseqdata/Homo_sapiens/featureCounts_out/coexp_new/merge/Homo_sapiens.spexp.pglnc_top20coexp.id.txt

#coexp基因作为输入，使用cytoscape 基于stringdb 建立PPI网络（默认参数,minimum required interaction score: 0.4），MCODE方法对网络分群，保留基因数量Top15 的cluster，分群富集
#物种ku
b <- 'org.Hs.eg.db'
bb <- "hsa"
#输入pglnc——TOP20共表达基因 cluster文件
a <- "~/PG/RNAseq/Homo_sapiens/addtarget_new/pglnc_top20coexp.csv"
#输出富集结果和plot
dd <- "/home/zhzhang/PG/RNAseq/Homo_sapiens/addtarget_new/Homo_sapiens.spexp.pglnc_top20coexp.cluster.enrich.txt"
ddd <- "/home/zhzhang/PG/RNAseq/plot/Homo_sapiens.spexp.pglnc_top20coexp.cluster.enrich.pdf"
#导入coexp基因 cluster文件
node <- read.csv(a)%>%
  select(1,2,4)%>%
  filter(Status=="Clustered")%>%
  select(3,1)
colnames(node) <- c("geneid","cluster")
tjclusternum <- group_by(node,cluster)%>%
  summarise(num=n())%>%
  top_n(15,num)%>%
  arrange(desc(num))%>%
  mutate(new_cluster=1:15)
clusterlist <- tjclusternum$cluster
newclusterlist <- tjclusternum$new_cluster
#循环富集
he <- data.frame()
for (i in 1:15) {
  fornode <- filter(node,cluster==clusterlist[i])%>%
    select(geneid)
  #GO富集分析
  parent_GO <- clusterProfiler::enrichGO(gene = fornode$geneid, OrgDb = b, keyType = "ENSEMBL", ont = 'BP', pAdjustMethod = 'fdr', pvalueCutoff = 1, qvalueCutoff = 1)
  if (is.null(parent_GO) == F) {
    parent_GO_result <- parent_GO@result%>%
      filter(p.adjust<0.05)
  }else {
    parent_GO_result <- data.frame()
  }
  #KEGG富集仅能识别entrze id，所以对基因ID进行转换
  ENTREZIDgenes <- clusterProfiler::bitr(fornode$geneid, fromType="ENSEMBL", toType="ENTREZID", OrgDb=b, drop = T)
  #KEGG富集分析
  GENEKEGG <- clusterProfiler::enrichKEGG(gene = ENTREZIDgenes$ENTREZID, keyType = 'kegg', organism = bb, pAdjustMethod = 'fdr', pvalueCutoff = 1, qvalueCutoff = 1) 
  if (is.null(GENEKEGG) == F) {
    parent_KEGG_result <- GENEKEGG@result%>%
      filter(p.adjust<0.05)
  }else {
    parent_KEGG_result <- data.frame()
  }
  #合并储存
  parent_enrich <- rbind(parent_GO_result,parent_KEGG_result)%>%
    separate(GeneRatio,c("no","all"),sep="/",remove = F)%>%
    mutate(all=as.numeric(all))%>%
    mutate(Generatio=Count/all)%>%
    arrange(Generatio)%>%
    select(-no,-all)%>%
    mutate(Description=Hmisc::capitalize(Description))%>%
    mutate(Description=do::Replace(data=Description,
                                   pattern=c("NcRNA:ncRNA","MRNA:mRNA",
                                             "RRNA:rRNA","PiRNA:piRNA",
                                             "SnRNA:snRNA")))%>%
    mutate(cluster=newclusterlist[i])
  #合并
  he <- rbind(he,parent_enrich)
}
#储存
data.table::fwrite(he,file =dd,
                   sep = '\t',row.names = F,quote = F,col.names = T)
#提取每个cluster count TOP3 通路
filter_enrich <- filter(he,Description!="Cellular process involved in reproduction in multicellular organism")%>%
  filter(Description!="Cilium or flagellum-dependent cell motility")%>%
  filter(Description!="Nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay")%>%
  filter(Description!="RNA destabilization")%>%
  mutate(Description=do::Replace(data=Description,
                                 pattern=c("NcRNA:ncRNA","MRNA:mRNA",
                                           "RRNA:rRNA","PiRNA:piRNA",
                                           "SnRNA:snRNA")))%>%
  filter(Count>1)%>%
  group_by(cluster)%>%
  top_n(3,Count)%>%
  top_n(3,-log10(p.adjust))%>%
  top_n(3,-log10(pvalue))
#plot（图大小：小鼠设为，人为12*15）
filter_enrich$Description[filter_enrich$Description=="Immune response-regulating cell surface receptor signaling pathway"] <- "Immune response-regulating cell\nsurface receptor signaling pathway"
filter_enrich$Description[filter_enrich$Description=="Regulation of G protein-coupled receptor signaling pathway"] <- "Regulation of G protein-coupled\nreceptor signaling pathway"

filter_enrich$cluster <- factor(as.character(filter_enrich$cluster),
                                levels = as.character(c(1:15)))
filter_enrich$Description <- factor(filter_enrich$Description,
                                    levels = distinct(data.frame("Description"=filter_enrich$Description),Description)$Description)
enrich <- ggplot(data = filter_enrich,aes(x=cluster,y=Description))+
  geom_point(aes(color=-log10(p.adjust),size=Count))+
  scale_color_gradient(low="#FF9569",high="#E92758")+
  theme_half_open()+
  labs(x = NULL, y =NULL)+
  theme(axis.text.x = element_text(angle =45))+
  theme(axis.text.x = element_text(vjust = 1,hjust=1))+
  theme(axis.title = element_text(size = 15),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),legend.text = element_text(size = 14)) +
  theme(panel.grid.major = element_line(colour = "gray95",
                                        size = 0.5), panel.grid.minor = element_line(colour = "gray95",
                                                                                     size = 0.5)) + theme(panel.grid.major = element_line(linetype = "longdash"),
                                                                                                          panel.grid.minor = element_line(linetype = "longdash"))
ggsave(ddd,
       enrich,width=11,height=17,dpi=1200,units="in",device='pdf',bg ="transparent")





###sup按照组织特异性富集top20靶基因
#不同组织特异性蛋白编码基因
q <- "~/PG/RNAseq/Homo_sapiens/Homo_sapiens.tissuewide1.pcgene.id.txt"
q <- "~/PG/RNAseq/Homo_sapiens/Homo_sapiens.tissuewide2.pcgene.id.txt"
q <- "~/PG/RNAseq/Homo_sapiens/Homo_sapiens.tissuewide3.pcgene.id.txt"
#导入pglnc top20共表达对
pglnc_pcgene_cordata_g20 <- read.delim("~/PG/RNAseq/Homo_sapiens/addtarget_new/Homo_sapiens.spexp.pglnc_pcgene.coexppair.top20.txt")

wide1_pcgene <- read.table(q, quote="\"", comment.char="")
colnames(wide1_pcgene) <- "targetid"
wide1_tar <- dplyr::inner_join(wide1_pcgene,select(pglnc_pcgene_cordata_g20,targetid),by="targetid")%>%
  distinct(targetid)
colnames(wide1_tar) <- "geneid"

#输出不同组织特异性pglnc靶基因富集结果
dd <- "/home/zhzhang/PG/RNAseq/Homo_sapiens/addtarget_new/Homo_sapiens.spexp.pglnc_top20coexp.filter1.enrich.txt"
dd <- "/home/zhzhang/PG/RNAseq/Homo_sapiens/addtarget_new/Homo_sapiens.spexp.pglnc_top20coexp.filter2.enrich.txt"
dd <- "/home/zhzhang/PG/RNAseq/Homo_sapiens/addtarget_new/Homo_sapiens.spexp.pglnc_top20coexp.filter3.enrich.txt"

b <- 'org.Hs.eg.db'
bb <- "hsa"

#GO富集分析
parent_GO <- clusterProfiler::enrichGO(gene = wide1_tar$geneid, OrgDb = b, keyType = "ENSEMBL", ont = 'BP', pAdjustMethod = 'fdr', pvalueCutoff = 1, qvalueCutoff = 1)
parent_GO_result <- parent_GO@result%>%
  filter(p.adjust<0.05)
#KEGG富集仅能识别entrze id，所以对基因ID进行转换
ENTREZIDgenes <- clusterProfiler::bitr(wide1_tar$geneid, fromType="ENSEMBL", toType="ENTREZID", OrgDb=b, drop = T)
#KEGG富集分析
GENEKEGG <- clusterProfiler::enrichKEGG(gene = ENTREZIDgenes$ENTREZID, keyType = 'kegg', organism = bb, pAdjustMethod = 'fdr', pvalueCutoff = 1, qvalueCutoff = 1) 
parent_KEGG_result <- GENEKEGG@result%>%
  filter(p.adjust<0.05)
#合并储存
parent_enrich <- rbind(parent_GO_result,parent_KEGG_result)%>%
  separate(GeneRatio,c("no","all"),sep="/",remove = F)%>%
  mutate(all=as.numeric(all))%>%
  mutate(Generatio=Count/all)%>%
  arrange(-Count)%>%
  select(-no,-all)%>%
  mutate(Description=Hmisc::capitalize(Description))%>%
  mutate(Description=do::Replace(data=Description,pattern=c("NcRNA:ncRNA","MRNA:mRNA","RRNA:rRNA","PiRNA:piRNA")))
data.table::fwrite(parent_enrich,file =dd,
                   sep = '\t',row.names = F,quote = F,col.names = T)



#导入不同组织特异性靶基因富集结果
filter1 <- read.delim("/home/zhzhang/PG/RNAseq/Homo_sapiens/addtarget_new/Homo_sapiens.spexp.pglnc_top20coexp.filter1.enrich.txt", 
                      header=T)
filter2 <- read.delim("/home/zhzhang/PG/RNAseq/Homo_sapiens/addtarget_new/Homo_sapiens.spexp.pglnc_top20coexp.filter2.enrich.txt", 
                      header=T)
filter3 <- read.delim("/home/zhzhang/PG/RNAseq/Homo_sapiens/addtarget_new/Homo_sapiens.spexp.pglnc_top20coexp.filter3.enrich.txt", 
                      header=T)
#提取基因数top10的通路画图展示
filter1_enrich <- filter(filter1,Description!="Cilium or flagellum-dependent cell motility")%>%
  filter(Description!="Cellular process involved in reproduction in multicellular organism")%>%
  top_n(10,Count)%>%
  top_n(10,-log10(p.adjust))%>%
  mutate(type="[Min, 10)")
filter2_enrich <- filter(filter2,Description!="Cellular process involved in reproduction in multicellular organism")%>%
  top_n(10,Count)%>%
  top_n(10,-log10(p.adjust))%>%
  mutate(type="[10, 20)")
filter3_enrich <- filter(filter3,Description!="Cellular process involved in reproduction in multicellular organism")%>%
  top_n(10,Count)%>%
  top_n(10,-log10(p.adjust))%>%
  mutate(type="[20, Max]")
parent_enrich <- rbind(filter1_enrich,filter2_enrich,filter3_enrich)
parent_enrich$type <- factor(parent_enrich$type,
                             levels = c("[Min, 10)","[10, 20)","[20, Max]"))
parent_enrich$Description <- factor(parent_enrich$Description,
                                    levels = distinct(select(parent_enrich,Description))$Description)
#plot
enrich <- ggplot(data = parent_enrich,aes(x=type,y=Description))+
  geom_point(aes(color=-log10(p.adjust),size=Count))+
  scale_color_gradient(low="#FF9569",high="#E92758")+
  theme_half_open()+
  labs(x = "Number of tissues", y =NULL)+
  theme(axis.text.x = element_text(angle =45))+
  theme(axis.text.x = element_text(vjust = 1,hjust=1))+
  theme(axis.title = element_text(size = 15),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),legend.text = element_text(size = 14))
ggsave("/home/zhzhang/PG/RNAseq/plot/Homo_sapiens.spexp.pglnc_top20coexp.filter123.enrich.pdf",
       enrich,width=10,height=8,dpi=1200,units="in",device='pdf',bg ="transparent")




```


##### 小鼠
```r
#分裂lncid（1000 line per file）
cat /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/Mus_musculus.spexp_pgdlncgene.id.txt /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/Mus_musculus.spexp_npgdlncgene.id.txt |split -l 1000 -d - /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/lncid/lncid
ls /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/lncid/ > /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/split.lncid.txt


#分裂pcgid（10 line per file）
split -l 10 -d /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/Mus_musculus.spexp_pcgene.id.txt /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/pcgid/pcgid
ls /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/pcgid/ > /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/split.pcgid.txt


#找到全部共表达的lnc-pcg基因对
#R
library(magrittr)
#生成参数列表
option_list <- list(optparse::make_option(
  opt_str ="--lncid",
  type = "character",
  default = NULL,
  help="splited spexp protein-coding geneid txt file"),
  optparse::make_option(
    opt_str ="--pcgid",
    type = "character",
    default = NULL,
    help="splited spexp protein-coding geneid txt file"),
  optparse::make_option(
    opt_str ="--outputone",
    type = "character",
    default = NULL,
    help="splited out"
  ),
  optparse::make_option(
    opt_str ="--outputtwo",
    type = "character",
    default = NULL,
    help="splited and filtered out"
  ))
#将参数列表传递给一个变量
args <- optparse::parse_args(optparse::OptionParser(option_list=option_list))
getgenepaircor <- function(a,b,c,d,e){
  #导入表达的lnc基因list（1000）
  lncid <- read.delim(b)
  colnames(lncid) <- "geneid"
  #导入表达的蛋白编码基因list（10）
  pcgid <- read.delim(c)
  colnames(pcgid) <- "geneid"
  #导入基因表达量TPM矩阵
  allsample_TPM <- data.table::fread(a)%>%
    data.frame()
  colnames(allsample_TPM)[1] <- "geneid"
  #lnc基因表达矩阵
  lncid_allsample_TPM <- dplyr::inner_join(allsample_TPM,lncid,by="geneid")%>%
    tibble::column_to_rownames("geneid")%>%
    t()%>%
    data.frame()
  #蛋白基因表达矩阵
  pcgid_allsample_TPM <- dplyr::inner_join(allsample_TPM,pcgid,by="geneid")%>%
    tibble::column_to_rownames("geneid")%>%
    t()%>%
    data.frame()
      #计算基因对的表达相关性
      cor <- psych::corr.test(x=lncid_allsample_TPM,
                              y=pcgid_allsample_TPM,
                              method="pearson",adjust="holm")
      perrson <- cor[["r"]]%>%
        data.frame()%>%
        tibble::rownames_to_column("geneid")%>%
        tidyr::gather("pcgid","cor",2:(nrow(pcgid)+1))
      pvalue <- cor[["p.adj"]]%>%
        data.frame()%>%
        tibble::rownames_to_column("geneid")%>%
        tidyr::gather("pcgid","pvalue",2:(nrow(pcgid)+1))%>%
        dplyr::select(pvalue)
      corandpadj <- cbind(perrson,pvalue)
  #保存
  data.table::fwrite(corandpadj,file =d,
                     sep = '\t',row.names = F,quote = F,col.names = F)
  #筛选
  filtercorlong <- dplyr::filter(corandpadj,pvalue<0.01 & abs(cor)>0.75)
  data.table::fwrite(filtercorlong,file =e,
                     sep = '\t',row.names = F,quote = F,col.names = F)
}
getgenepaircor(a="/home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/cbindsample_TPM.txt",
               b=args$lncid,
               c=args$pcgid,
               d=args$outputone,
               e=args$outputtwo)

#parallel 并行执行R脚本
cat "/home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/split.lncid.txt"|while read i
do
echo "${i}"
cat "/home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/split.pcgid.txt" | parallel -j 120 --verbose Rscript "/home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/sh/mm.R" --lncid /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/lncid/${i} --pcgid /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/pcgid/{} --outputone /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/out1/${i}.{}.out --outputtwo /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/out2/${i}.{}.out
done


#合并cor并行结果
cat /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/out2/* > /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/merge/Mus_musculus.spexp.lnc_pcgene.cordata.filter.txt
#获取全部lnc的coexp蛋白编码基因id（作为证明pglnc通路更focus在精子发生时，富集的bgd）
awk '{print $2}' "/home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/merge/Mus_musculus.spexp.lnc_pcgene.cordata.filter.txt"|sort|uniq > /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/merge/Mus_musculus.spexp.lnc_coexp_pcg.bgd.id.txt
#提取pglnc-pcg共表达对
grep -w -Ff "/home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/Mus_musculus.spexp_pgdlncgene.id.txt" "/home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/merge/Mus_musculus.spexp.lnc_pcgene.cordata.filter.txt" > "/home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/merge/Mus_musculus.spexp.pglnc_pcgene.cordata.filter.txt"
#all pglnc coexp pcgid
awk '{print $2}' /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/merge/Mus_musculus.spexp.pglnc_pcgene.cordata.filter.txt|sort|uniq > /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/merge/Mus_musculus.spexp.pglnc_coexp.id.txt 
#提取npglnc-pcg共表达对
grep -v -w -Ff "/home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/Mus_musculus.spexp_pgdlncgene.id.txt" "/home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/merge/Mus_musculus.spexp.lnc_pcgene.cordata.filter.txt" > "/home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/merge/Mus_musculus.spexp.npglnc_pcgene.cordata.filter.txt"
#all npglnc coexp pcgid
awk '{print $2}' /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/merge/Mus_musculus.spexp.npglnc_pcgene.cordata.filter.txt|sort|uniq > /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/merge/Mus_musculus.spexp.npglnc_coexp.id.txt 




###证明pglnc通路更focus在精子发生
#导入两类lnc的靶基因
pgcoexp <- read.table("/home/zhzhang/PG/RNAseq/Mus_musculus/addtarget_new/Mus_musculus.spexp.pglnc_coexp.id.txt", quote="\"", comment.char="")
colnames(pgcoexp) <- "geneid"
npgcoexp <- read.table("/home/zhzhang/PG/RNAseq/Mus_musculus/addtarget_new/Mus_musculus.spexp.npglnc_coexp.id.txt", quote="\"", comment.char="")
colnames(npgcoexp) <- "geneid"
#导入背景基因集
bgd <- read.table("/home/zhzhang/PG/RNAseq/Mus_musculus/addtarget_new/Mus_musculus.spexp.lnc_coexp_pcg.bgd.id.txt",
                  quote="\"", comment.char="")
colnames(bgd) <- "geneid"
#pglnc all靶基因富集信息
pgcoexp_GO <- clusterProfiler::enrichGO(gene = pgcoexp$geneid, universe=bgd$geneid,OrgDb = 'org.Mm.eg.db', keyType = "ENSEMBL", ont = 'BP', pAdjustMethod = 'fdr', pvalueCutoff = 0.05, qvalueCutoff = 1)
pgcoexp_GO_result <- pgcoexp_GO@result
pgcoexp_GO_result <- pgcoexp_GO_result[which(pgcoexp_GO_result$Description=="germ cell development"|
                                               pgcoexp_GO_result$Description=="meiotic cell cycle"|
                                               pgcoexp_GO_result$Description=="meiotic cell cycle process"|
                                               pgcoexp_GO_result$Description=="meiotic nuclear division"|
                                               pgcoexp_GO_result$Description=="cilium organization"|
                                               pgcoexp_GO_result$Description=="cilium assembly"|
                                               pgcoexp_GO_result$Description=="cilium movement"|
                                               pgcoexp_GO_result$Description=="cilium-dependent cell motility"|
                                               pgcoexp_GO_result$Description=="spermatid differentiation"|
                                               pgcoexp_GO_result$Description=="spermatid development"),]

pgcoexp_GO_result <- separate(pgcoexp_GO_result,GeneRatio,c("no","all"),sep="/",remove = F)%>%
  mutate(all=as.numeric(all))%>%
  mutate(Generatio=Count/all)%>%
  select(-no,-all)%>%
  separate(BgRatio,c("BgRatio1","BgRatio2"),sep="/",remove = F)%>%
  mutate(BgRatio1=as.numeric(BgRatio1))%>%
  mutate(BgRatio2=as.numeric(BgRatio2))%>%
  mutate(Bgratio=BgRatio1/BgRatio2)%>%
  select(-BgRatio1,-BgRatio2)%>%
  mutate(Description=Hmisc::capitalize(Description))%>%
  mutate(ES=Generatio/Bgratio)%>%
  mutate(type="Pseudogene-derived lncRNA")
#npglnc all靶基因富集信息
npg_GO <- clusterProfiler::enrichGO(gene = npgcoexp$geneid, universe=bgd$geneid,OrgDb = 'org.Mm.eg.db', keyType = "ENSEMBL", ont = 'BP', pAdjustMethod = 'fdr', pvalueCutoff = 0.05, qvalueCutoff = 1)
npg_GO_result <- npg_GO@result
npg_GO_result <- npg_GO_result[which(npg_GO_result$Description=="germ cell development"|
                                       npg_GO_result$Description=="meiotic cell cycle"|
                                       npg_GO_result$Description=="meiotic cell cycle process"|
                                       npg_GO_result$Description=="meiotic nuclear division"|
                                       npg_GO_result$Description=="cilium organization"|
                                       npg_GO_result$Description=="cilium assembly"|
                                       npg_GO_result$Description=="cilium movement"|
                                       npg_GO_result$Description=="cilium-dependent cell motility"|
                                       npg_GO_result$Description=="spermatid differentiation"|
                                       npg_GO_result$Description=="spermatid development"),]
npg_GO_result <- separate(npg_GO_result,GeneRatio,c("no","all"),sep="/",remove = F)%>%
  mutate(all=as.numeric(all))%>%
  mutate(Generatio=Count/all)%>%
  select(-no,-all)%>%
  separate(BgRatio,c("BgRatio1","BgRatio2"),sep="/",remove = F)%>%
  mutate(BgRatio1=as.numeric(BgRatio1))%>%
  mutate(BgRatio2=as.numeric(BgRatio2))%>%
  mutate(Bgratio=BgRatio1/BgRatio2)%>%
  select(-BgRatio1,-BgRatio2)%>%
  mutate(Description=Hmisc::capitalize(Description))%>%
  mutate(ES=Generatio/Bgratio)%>%
  mutate(type="Non-pseudogene-derived lncRNA")
#hebing
lnche <- rbind(pgcoexp_GO_result,npg_GO_result)
lnche$type <- factor(lnche$type,levels=c("Non-pseudogene-derived lncRNA",
                                         "Pseudogene-derived lncRNA"))
lnche$Description <- factor(lnche$Description,levels=c("Spermatid development",
                                                       "Spermatid differentiation",
                                                       "Cilium-dependent cell motility",
                                                       "Cilium movement",
                                                       "Cilium assembly",
                                                       "Cilium organization",
                                                       "Meiotic nuclear division",
                                                       "Meiotic cell cycle process",
                                                       "Meiotic cell cycle",
                                                       "Germ cell development"))
#储存
data.table::fwrite(lnche,file ="/home/zhzhang/PG/RNAseq/Mus_musculus/addtarget_new/Mus_musculus.spexp.allpglnc_npglnc_coexpgcgene.bgdenrich.txt",
                   sep = '\t',row.names = F,quote = F,col.names = T)
#富集dotplot
lnche <- mutate(lnche,anno="N.S.")
lnche$anno[lnche$p.adjust < 0.05] <- "*"
lnche$anno[lnche$p.adjust < 0.01] <- "**"
lnche$anno[lnche$p.adjust < 0.001] <- "***"
pp1 <- ggplot(data = lnche,aes(x=type,y=Description))+
  geom_point(aes(fill=-log10(p.adjust),size=ES),shape=21,color="black",
             position =position_nudge(x=0.15))+
  geom_text(aes(label=anno),nudge_x =-0.15 )+
  geom_hline(yintercept=c(1.5:9.5),color="gray")+
  geom_vline(xintercept=c(1.5),color="gray")+
  scale_fill_gradient(low="white",high="#C12039")+
  scale_x_discrete(labels=c("Non-pseudogene-\nderived lncRNA",
                            "Pseudogene-derived\nlncRNA"))+
  theme_half_open()+
  scale_y_discrete(position = "right")+
  labs(x = NULL, y =NULL,size="Enrichment score")+
  theme(axis.text.x = element_text(angle =45))+
  theme(axis.text.x = element_text(vjust = 1,hjust=1))+
  theme(axis.title = element_text(size = 15),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),legend.text = element_text(size = 14)) +
  theme(panel.background = element_rect(colour = "black",size=1)) +
  theme(legend.position = "left")
ggsave("/home/zhzhang/PG/RNAseq/plot/Mus_musculus.spexp.allpglnc_npglnc_coexpgcgene.bgdenrich.pdf", 
       pp1,width = 7, height = 6,dpi=1200, units = "in", device='pdf',bg = "transparent")












#TOP20共表达基因用于构建PPI网络
#提取每个pglnc top20共表达pcg
a <- "/home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/merge/Mus_musculus.spexp.pglnc_pcgene.cordata.filter.txt"
c <- "/home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/merge/Mus_musculus.spexp.pglnc_pcgene.coexppair.top20.txt"
#导入基因对cor
pglnc_pcgene_cordata <- read.delim(a, header=FALSE)
pglnc_pcgene_cordata_g20 <- group_by(pglnc_pcgene_cordata,V1)%>%
  top_n(20,V3)
#储存筛选后的pglnc及其靶基因
colnames(pglnc_pcgene_cordata_g20) <- c("lncgid","targetid","cor","padj")
data.table::fwrite(pglnc_pcgene_cordata_g20,
                   file =c,sep = '\t',row.names = F,quote = F,col.names = T)
#pglnc top20共表达pcgid
tail -n +2 /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/merge/Mus_musculus.spexp.pglnc_pcgene.coexppair.top20.txt|awk '{print $2}'|sort|uniq > /home/zhzhang/PG/RNAseqdata/Mus_musculus/featureCounts_out/coexp_new/merge/Mus_musculus.spexp.pglnc_top20coexp.id.txt

#coexp基因作为输入，使用cytoscape 基于stringdb 建立PPI网络（默认参数,minimum required interaction score: 0.3），MCODE方法对网络分群，保留基因数量Top15 的cluster，分群富集
#物种ku
b <- 'org.Mm.eg.db'
bb <- "mmu"
#输入pglnc——TOP20共表达基因 cluster文件
a <- "/home/zhzhang/PG/RNAseq/Mus_musculus/addtarget_new/mmpglnc_top20coexp.csv"
#输出富集结果和plot
dd <- "/home/zhzhang/PG/RNAseq/Mus_musculus/addtarget_new/Mus_musculus.spexp.pglnc_top20coexp.cluster.enrich.txt"
ddd <- "/home/zhzhang/PG/RNAseq/plot/Mus_musculus.spexp.pglnc_top20coexp.cluster.enrich.pdf"
#导入coexp基因 cluster文件
node <- read.csv(a)%>%
  select(1,2,4)%>%
  filter(Status!="Unclustered")%>%
  select(3,1)
colnames(node) <- c("geneid","cluster")
tjclusternum <- group_by(node,cluster)%>%
  summarise(num=n())%>%
  top_n(15,num)%>%
  arrange(desc(num))%>%
  mutate(new_cluster=1:15)
clusterlist <- tjclusternum$cluster
newclusterlist <- tjclusternum$new_cluster
#循环富集
he <- data.frame()
for (i in 1:15) {
  fornode <- filter(node,cluster==clusterlist[i])%>%
    select(geneid)
  #GO富集分析
  parent_GO <- clusterProfiler::enrichGO(gene = fornode$geneid, OrgDb = b, keyType = "ENSEMBL", ont = 'BP', pAdjustMethod = 'fdr', pvalueCutoff = 1, qvalueCutoff = 1)
  if (is.null(parent_GO) == F) {
    parent_GO_result <- parent_GO@result%>%
      filter(p.adjust<0.05)
  }else {
    parent_GO_result <- data.frame()
  }
  #KEGG富集仅能识别entrze id，所以对基因ID进行转换
  ENTREZIDgenes <- clusterProfiler::bitr(fornode$geneid, fromType="ENSEMBL", toType="ENTREZID", OrgDb=b, drop = T)
  #KEGG富集分析
  GENEKEGG <- clusterProfiler::enrichKEGG(gene = ENTREZIDgenes$ENTREZID, keyType = 'kegg', organism = bb, pAdjustMethod = 'fdr', pvalueCutoff = 1, qvalueCutoff = 1) 
  if (is.null(GENEKEGG) == F) {
    parent_KEGG_result <- GENEKEGG@result%>%
      filter(p.adjust<0.05)
  }else {
    parent_KEGG_result <- data.frame()
  }
  #合并储存
  parent_enrich <- rbind(parent_GO_result,parent_KEGG_result)%>%
    separate(GeneRatio,c("no","all"),sep="/",remove = F)%>%
    mutate(all=as.numeric(all))%>%
    mutate(Generatio=Count/all)%>%
    arrange(Generatio)%>%
    select(-no,-all)%>%
    mutate(Description=Hmisc::capitalize(Description))%>%
    mutate(Description=do::Replace(data=Description,
                                   pattern=c("NcRNA:ncRNA","MRNA:mRNA",
                                             "RRNA:rRNA","PiRNA:piRNA",
                                             "SnRNA:snRNA")))%>%
    mutate(cluster=newclusterlist[i])
  #合并
  he <- rbind(he,parent_enrich)
}
#储存
data.table::fwrite(he,file =dd,
                   sep = '\t',row.names = F,quote = F,col.names = T)
#提取每个cluster count TOP5 通路
filter_enrich <- filter(he,Description!="Cellular process involved in reproduction in multicellular organism")%>%
  filter(Description!="Cilium or flagellum-dependent cell motility")%>%
  filter(Description!="Nuclear-transcribed mRNA catabolic process, deadenylation-dependent decay")%>%
  filter(Description!="RNA destabilization")%>%
  mutate(Description=do::Replace(data=Description,
                                 pattern=c("NcRNA:ncRNA","MRNA:mRNA",
                                           "RRNA:rRNA","PiRNA:piRNA",
                                           "SnRNA:snRNA")))%>%
  filter(Count>1)%>%
  group_by(cluster)%>%
  top_n(3,Count)%>%
  top_n(3,-log10(p.adjust))%>%
  top_n(3,-log10(pvalue))%>%
  separate(Description,c("Description","no"),sep=" - ")%>%
  select(-no)
#plot（图大小：小鼠设为，人为12*15）
filter_enrich$cluster <- factor(as.character(filter_enrich$cluster),
                                levels = as.character(c(1:15)))
filter_enrich$Description <- factor(filter_enrich$Description,
                                    levels = distinct(data.frame("Description"=filter_enrich$Description),Description)$Description)
enrich <- ggplot(data = filter_enrich,aes(x=cluster,y=Description))+
  geom_point(aes(color=-log10(p.adjust),size=Count))+
  scale_color_gradient(low="#FF9569",high="#E92758")+
  theme_half_open()+
  labs(x = NULL, y =NULL)+
  theme(axis.text.x = element_text(angle =45))+
  theme(axis.text.x = element_text(vjust = 1,hjust=1))+
  theme(axis.title = element_text(size = 15),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),legend.text = element_text(size = 14)) +
  theme(panel.grid.major = element_line(colour = "gray95",
                                        size = 0.5), panel.grid.minor = element_line(colour = "gray95",
                                                                                     size = 0.5)) + theme(panel.grid.major = element_line(linetype = "longdash"),
                                                                                                          panel.grid.minor = element_line(linetype = "longdash"))
ggsave(ddd,
       enrich,width=13,height=17,dpi=1200,units="in",device='pdf',bg ="transparent")



```

