### FIG6 
### lncRNA scrna
##### 1.数据下载
```r
#下sra数据
https://www.ebi.ac.uk/ena/browser/view/PRJNA757777

#创建参考基因组
cd /home/zhzhang/PG/scRNAseq/crref/
cellranger mkref --genome=grch38_108 --fasta=/home/zhzhang/PG/refgenome/Homo_sapiens.GRCh38.dna.chr.fa --genes=/home/zhzhang/PG/RNAseqdata/newGTF/Homo_sapiens.GRCh38.108.chr.rmpg.novellncRNA.gtf --nthreads=64 --memgb=254

#SRA生成fastq，压缩，以及标准命名
ls /home/zhzhang/PG/scRNAseq/sra/ > /home/zhzhang/PG/scRNAseq/samplename.txt
cat /home/zhzhang/PG/scRNAseq/samplename.txt|while read i
do
fasterq-dump -e 64 -m 2048GB --split-3 -O /home/zhzhang/PG/scRNAseq/fastq/${i}/ -o ${i} /home/zhzhang/PG/scRNAseq/sra/${i}
pigz -p 64 /home/zhzhang/PG/scRNAseq/fastq/${i}/${i}_1.fastq
pigz -p 64 /home/zhzhang/PG/scRNAseq/fastq/${i}/${i}_2.fastq
mv /home/zhzhang/PG/scRNAseq/fastq/${i}/${i}_1.fastq /home/zhzhang/PG/scRNAseq/fastq/${i}/${i}_S1_L001_R1_001.fastq.gz
mv /home/zhzhang/PG/scRNAseq/fastq/${i}/${i}_2.fastq /home/zhzhang/PG/scRNAseq/fastq/${i}/${i}_S1_L001_R2_001.fastq.gz
done




#产生fastq后，根据reads长度判断reads类型，然后按照下述格式规范命名
[Sample Name]_S1_L00[Lane Number]_[Read Type]_001.fastq.gz
seqkit stats fq文件路径 -T -j 10



```


##### 2.产生表达矩阵
```r
cd /home/zhzhang/PG/scRNAseq/count/
cat /home/zhzhang/PG/scRNAseq/samplename.txt|while read i
do
cellranger count --id=${i} --transcriptome=/home/zhzhang/PG/scRNAseq/crref/grch38_108/ --fastqs=/home/zhzhang/PG/scRNAseq/fastq/${i}/ --sample=${i} --nosecondary --localcores=64 --localmem=1024
done

#cellbender去除背景RNA污染，去除背景barcode
cat /home/zhzhang/PG/scRNAseq/samplename.txt|while read i
do
cellbender remove-background --input /home/zhzhang/PG/scRNAseq/count/${i}/outs/raw_feature_bc_matrix.h5 --output /home/zhzhang/PG/scRNAseq/cellbender/${i}/${i}_feature_bc_matrix.h5 --cpu-threads 128
done



```


##### 3.细胞聚类
```r
library(dplyr)
library(Seurat)
library(patchwork)



# Load the dataset
O1 <- Read10X(data.dir = "/home/zhzhang/PG/scRNAseq/count/O1/outs/filtered_feature_bc_matrix/")
O2 <- Read10X(data.dir = "/home/zhzhang/PG/scRNAseq/count/O2/outs/filtered_feature_bc_matrix/")
O3 <- Read10X(data.dir = "/home/zhzhang/PG/scRNAseq/count/O3/outs/filtered_feature_bc_matrix/")
O4 <- Read10X(data.dir = "/home/zhzhang/PG/scRNAseq/count/O4/outs/filtered_feature_bc_matrix/")
O5 <- Read10X(data.dir = "/home/zhzhang/PG/scRNAseq/count/O5/outs/filtered_feature_bc_matrix/")
O6 <- Read10X(data.dir = "/home/zhzhang/PG/scRNAseq/count/O6/outs/filtered_feature_bc_matrix/")
O7 <- Read10X(data.dir = "/home/zhzhang/PG/scRNAseq/count/O7/outs/filtered_feature_bc_matrix/")
O8 <- Read10X(data.dir = "/home/zhzhang/PG/scRNAseq/count/O8/outs/filtered_feature_bc_matrix/")
Y1 <- Read10X(data.dir = "/home/zhzhang/PG/scRNAseq/count/Y1/outs/filtered_feature_bc_matrix/")
Y2 <- Read10X(data.dir = "/home/zhzhang/PG/scRNAseq/count/Y2/outs/filtered_feature_bc_matrix/")
Y3 <- Read10X(data.dir = "/home/zhzhang/PG/scRNAseq/count/Y3/outs/filtered_feature_bc_matrix/")
Y4 <- Read10X(data.dir = "/home/zhzhang/PG/scRNAseq/count/Y4/outs/filtered_feature_bc_matrix/")
#创建对象
O1 <- CreateSeuratObject(counts = O1, project = "O1", min.cells = 3, min.features = 200)
O2 <- CreateSeuratObject(counts = O2, project = "O2", min.cells = 3, min.features = 200)
O3 <- CreateSeuratObject(counts = O3, project = "O3", min.cells = 3, min.features = 200)
O4 <- CreateSeuratObject(counts = O4, project = "O4", min.cells = 3, min.features = 200)
O5 <- CreateSeuratObject(counts = O5, project = "O5", min.cells = 3, min.features = 200)
O6 <- CreateSeuratObject(counts = O6, project = "O6", min.cells = 3, min.features = 200)
O7 <- CreateSeuratObject(counts = O7, project = "O7", min.cells = 3, min.features = 200)
O8 <- CreateSeuratObject(counts = O8, project = "O8", min.cells = 3, min.features = 200)
Y1 <- CreateSeuratObject(counts = Y1, project = "Y1", min.cells = 3, min.features = 200)
Y2 <- CreateSeuratObject(counts = Y2, project = "Y2", min.cells = 3, min.features = 200)
Y3 <- CreateSeuratObject(counts = Y3, project = "Y3", min.cells = 3, min.features = 200)
Y4 <- CreateSeuratObject(counts = Y4, project = "Y4", min.cells = 3, min.features = 200)

#merge
ALLsample <- merge(O1,y = c(O2,O3,O4,O5,O6,O7,O8,Y1,Y2,Y3,Y4), add.cell.ids = c("O1","O2","O3","O4","O5","O6","O7","O8","Y1","Y2","Y3","Y4"), project = "ALLsample")

#细胞质控
#VlnPlot(ALLsample, features = c("nFeature_RNA", "nCount_RNA"), ncol =2, group.by = "orig.ident", pt.size = 0)
ALLsample <- subset(ALLsample, subset = nFeature_RNA > 800 )
#标准化归一化
ALLsample <- SCTransform(ALLsample,variable.features.n=5000)
#降维
ALLsample <- RunPCA(ALLsample,npcs = 50)
#DimPlot(object = ALLsample, dims=c(1,2),reduction = "pca")
#ElbowPlot(ALLsample, ndims = 50, reduction = "pca")
#聚类
ALLsample <- FindNeighbors(ALLsample, dims = 1:40)
ALLsample <- FindClusters(ALLsample, resolution = 0.4)
#clustree(ALLsample)
#DimPlot(object = ALLsample, reduction = "pca",dim=c(1,2),label = TRUE,group.by="orig.ident")
#可视化降维
ALLsample <- RunUMAP(ALLsample, dims = 1:40)
#DimPlot(object = ALLsample, reduction = "umap",dim=c(1,2),group.by="seurat_clusters",label = TRUE)
#保存Seurat对象
saveRDS(ALLsample, file = "/home/zhzhang/PG/scRNAseq/seurat/cluster_ALLsample.rds")


```
```r
library(dplyr)
library(Seurat)
library(patchwork)


ALLsample <- readRDS("~/PG/scRNAseq/seurat/cluster_ALLsample.rds")
#去除双细胞
library(DoubletFinder)
#寻找最优pK值
sweep.res.list <- paramSweep_v3(ALLsample, PCs = 1:40, sct = T, num.cores=96)
sweep.stats <- summarizeSweep(sweep.res.list, GT = FALSE)
bcmvn <- find.pK(sweep.stats) #可以看到最佳参数的点
## 所以最佳的参数是：
mpK<-as.numeric(as.vector(bcmvn$pK[which.max(bcmvn$BCmetric)]))
#全部doublets比例
DoubletRate <- 0.039
#估计同源双细胞比例
homotypic.prop <- modelHomotypic(ALLsample@meta.data[["seurat_clusters"]])
#双细胞数量
nExp_poi <- round(DoubletRate*ncol(ALLsample)) 
#异源双细胞数量
nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
#检测
ALLsample <- doubletFinder_v3(ALLsample, PCs = 1:40, pN = 0.25, pK = mpK, nExp = nExp_poi.adj, sct = T)
saveRDS(ALLsample, file = "/home/zhzhang/PG/scRNAseq/seurat/doudetec_ALLsample.rds")
grep("DF",colnames(ALLsample@meta.data),value = T)

```
```r
#SCT findmarker
library(dplyr)
library(Seurat)
library(patchwork)
ALLsample <- readRDS("~/PG/scRNAseq/seurat/doudetec_ALLsample.rds")
ALLsample <- subset(ALLsample, subset = DF.classifications_0.25_0.01_1969 == "Singlet" )
ALLsample <- PrepSCTFindMarkers(ALLsample, assay = "SCT")
ALLsample.markers <- FindAllMarkers(object = ALLsample,assay="SCT",only.pos = TRUE, min.pct = 0.25, logfc.threshold = 1)
saveRDS(ALLsample, file = "/home/zhzhang/PG/scRNAseq/seurat/SCTcorrect_ALLsample.rds")
data.table::fwrite(ALLsample.markers,file ="/home/zhzhang/PG/scRNAseq/seurat/ALLsample.allcluster.markergene.txt",sep = '\t',row.names = F,quote = F,col.names = T)


#RNA findmarker
library(dplyr)
library(Seurat)
library(patchwork)
ALLsample <- readRDS("~/PG/scRNAseq/seurat/doudetec_ALLsample.rds")
ALLsample <- subset(ALLsample, subset = DF.classifications_0.25_0.01_1969 == "Singlet" )
ALLsample <- NormalizeData(ALLsample, assay ="RNA",normalization.method = "LogNormalize", scale.factor = 10000)
ALLsample.markers <- FindAllMarkers(object = ALLsample,assay="RNA",only.pos = TRUE, min.pct = 0.25, logfc.threshold = 1,return.thresh = 0.01)
saveRDS(ALLsample, file = "/home/zhzhang/PG/scRNAseq/seurat/RNAnormal_ALLsample.rds")
data.table::fwrite(ALLsample.markers,file ="/home/zhzhang/PG/scRNAseq/seurat/ALLsample.allcluster.markergene.byRNA.txt",sep = '\t',row.names = F,quote = F,col.names = T)
```




##### 4.细胞注释
```r
scp -P 22 zhzhang@211.69.141.147:/home/zhzhang/PG/scRNAseq/seurat/ALLsample.allcluster.markergene.txt /home/zhzhang/PG/scRNAseq/seurat/
scp -P 22 zhzhang@211.69.141.147:/home/zhzhang/PG/scRNAseq/seurat/ALLsample.allcluster.markergene.byRNA.txt /home/zhzhang/PG/scRNAseq/seurat/
scp -P 22 zhzhang@211.69.141.147:/home/zhzhang/PG/scRNAseq/seurat/RNAnormal_ALLsample.rds /home/zhzhang/PG/scRNAseq/seurat/

library(dplyr)
library(Seurat)
library(patchwork)
#导入R
ALLsample <- readRDS("~/PG/scRNAseq/seurat/RNAnormal_ALLsample.rds")
DimPlot(object = ALLsample, reduction = "umap",group.by="seurat_clusters",label = TRUE)
#手工注释细胞类型
cluster2celltype <- c("0"="EC",
                      "1"="PMC", 
                      "2"="Elongating",
                      "3"= "VSMC", 
                      "4"= "Leydig", 
                      "5"= "Early SPC",
                      "6"= "RS", 
                      "7"= "PMC", 
                      "8"= "Diff.SPG",
                      "9"= "Elongated",
                      "10"= "Late SPC",
                      "11"= "SSC",
                      "12"= "Macrophages",
                      "13"= "RS",
                      "14"= "Elongating",
                      "15"= "Sertoli",
                      "16"= "Elongated",
                      "17"= "Sertoli",
                      "18"= "Sertoli",
                      "19"= "EC",
                      "20"= "Late SPC",
                      "21"= "Sertoli",
                      "22"= "EC",
                      "23"= "T-cell",
                      "24"= "T-cell"
)
ALLsample@meta.data[['cell_type']] <- unname(cluster2celltype[ALLsample@meta.data$seurat_clusters])
#设置每个细胞分类标识的变量名
Idents(ALLsample) <- "cell_type"
#plot
ALLsample@meta.data[['cell_type']] <- factor(ALLsample@meta.data[['cell_type']],
                                             levels =c("SSC","Diff.SPG","Early SPC","Late SPC",
                                                       "RS","Elongating","Elongated",
                                                       "Sertoli","Leydig","PMC","VSMC","EC",
                                                       "Macrophages","T-cell"))
umapcell <- DimPlot(object = ALLsample, reduction = "umap",group.by="cell_type",label = TRUE)+
  scale_color_manual(values = c("#2A6434","#50A15C","#7DBB78",
                                "#ACD18E","#A6C9DD","#619BC0",
                                "#337CA5","#E0DC95","#984E27",
                                "#B86577","#E2A0A4","#B58F7A",
                                "#BFB5D7","#D3A665"))+
  theme(axis.title = element_text(size = 15),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 15))+
  labs(title = NULL,x="UMAP-1",y="UMAP-2")+
  theme(legend.text = element_text(size = 20),legend.key.size = unit(25,"pt"))+
  guides(color=guide_legend(override.aes = list(size=5)))
ggsave("/home/zhzhang/PG/scRNAseq/seurat/Humantestis.celltype_umap.png", 
       umapcell,width = 9, height = 6.6,dpi=1200, units = "in", device='png',bg = "transparent")
#储存
saveRDS(ALLsample, file = "/home/zhzhang/PG/scRNAseq/seurat/celltype_ALLsample.rds")




```


##### \[5.\] 各cell type三类基因表达比例
```r
#单细胞基因ID和基因名对照表
tail -n +2 /home/zhzhang/PG/scRNAseq/crref/grch38_108/star/geneInfo.tab|awk '{print $1"\t"$2}' > /home/zhzhang/PG/scRNAseq/crref/gene.ID.name.human.txt


#基因名genename-基因类型对照表
#R
#导入单细胞基因ID和基因名对照表
geneIDname <- read.delim("~/PG/scRNAseq/gene.ID.name.human.txt", header=FALSE)
colnames(geneIDname) <- c("geneid","genename")
#导入基因ID分类
geneid_class <- read.delim("~/PG/RNAseq/Homo_sapiens/Homo_sapiens.geneid_class.txt")
#合并产生基因名-基因类型对照表
geneid_he <- filter(geneid_class,type!="Interference lncRNA")%>%
  left_join(geneIDname,by="geneid")%>%
  select(3,2)%>%
  mutate(genename=stringr::str_replace_all(genename,"_","-"))%>%#因为单细胞seurat包会把基因name中的_替换成-，所以这里与seurat保持一致
  distinct(genename,.keep_all = T)
#储存
data.table::fwrite(geneid_he,file ="/home/zhzhang/PG/scRNAseq/genename.type.human.txt",sep = '\t',row.names = F,quote = F,col.names = T)


```
```r
#生科院服务器，计算每个celltype细胞群中，每个基因的阳性细胞比例pct1
library(dplyr)
library(Seurat)
library(patchwork)
library(future)

ALLsample <- readRDS("/home/zhzhang/PG/scRNAseq/seurat/RNAnormal_ALLsample.rds")
#手工注释细胞类型
cluster2celltype <- c("0"="EC",
                      "1"="PMC", 
                      "2"="Elongating", 
                      "3"= "VSMC", 
                      "4"= "Leydig", 
                      "5"= "Early SPC",
                      "6"= "RS", 
                      "7"= "PMC", 
                      "8"= "Diff.SPG",
                      "9"= "Elongated",
                      "10"= "Late SPC",
                      "11"= "SSC",
                      "12"= "Macrophages",
                      "13"= "RS",
                      "14"= "Elongating",
                      "15"= "Sertoli",
                      "16"= "Elongated",
                      "17"= "Sertoli",
                      "18"= "Sertoli",
                      "19"= "EC",
                      "20"= "Late SPC",
                      "21"= "Sertoli",
                      "22"= "EC",
                      "23"= "T-cell",
                      "24"= "T-cell")
ALLsample@meta.data[['cell_type']] <- unname(cluster2celltype[ALLsample@meta.data$seurat_clusters])
#设置每个细胞分类标识的变量名
Idents(ALLsample) <- "cell_type"

plan("multiprocess", workers = 128)
options(future.globals.maxSize= 2500000000000)
ALLsample.markers <- FindAllMarkers(object = ALLsample,assay="RNA",min.pct=0,logfc.threshold=0,only.pos = FALSE,return.thresh=1)


data.table::fwrite(ALLsample.markers,file ="/home/zhzhang/PG/scRNAseq/seurat/ALLsample.allcelltype.markergene.byRNA.txt",sep = '\t',row.names = F,quote = F,col.names = T)



#
scp -P 22 zhzhang@211.69.141.147:/home/zhzhang/PG/scRNAseq/seurat/ALLsample.allcelltype.markergene.byRNA.txt /home/zhzhang/PG/scRNAseq/seurat/

```
```r
#对比每类细胞中三类基因的表达比例（在某细胞群中阳性细胞比例>=0.01即为在该细胞群中表达）
#导入基因name分类
genename_class <- read.delim("/home/zhzhang/PG/scRNAseq/genename.type.human.txt")
#导入全部基因在各celltype的阳性细胞比例，并添加基因分类
allcelltype_genepct <- read.delim("~/PG/scRNAseq/seurat/ALLsample.allcelltype.markergene.byRNA.txt")%>%
  select(6,7,3)
colnames(allcelltype_genepct) <- c("celltype","genename","pct.1")
allcelltype_genepct <- left_join(allcelltype_genepct,genename_class,by="genename")
allcelltype_genepct <- allcelltype_genepct[is.na(allcelltype_genepct$type)==F,]
#基因在某细胞群中阳性细胞比例>=0.01即为在该细胞群中表达(后续储存表达的基因列表)
allcelltype_genepct <- mutate(allcelltype_genepct,exp=0)
allcelltype_genepct$exp[allcelltype_genepct$pct.1>=0.01] <- 1
#tj三类基因总数
genenum <- group_by(mutate(genename_class,num=1),type)%>%
  summarise(allnum=sum(num))
#统计三类基因表达比例
celltype_expratio <- group_by(allcelltype_genepct,celltype,type)%>%
  summarise(expnum=sum(exp))%>%
  left_join(genenum,by="type")%>%
  mutate(expratio=expnum/allnum)
celltype_expratio$type <- factor(celltype_expratio$type,levels = c("Protein-coding","Pseudogene-derived lncRNA","Non-pseudogene-derived lncRNA"))
celltype_expratio$celltype <- factor(celltype_expratio$celltype,
                                     levels = c("SSC","Diff.SPG","Early SPC","Late SPC",
                                                "RS","Elongating","Elongated",
                                                "Sertoli","Leydig","PMC","VSMC","EC",
                                                "Macrophages","T-cell"))
data.table::fwrite(celltype_expratio,file ="/home/zhzhang/PG/scRNAseq/1expratio/Humantestis.lnc_celltype_exp_percent.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#验证人类睾丸每个celltype中pglnc表达比例高于npglnc表达比例的显著性（fishertest）
celltype_expratio <- mutate(celltype_expratio,nonum=allnum-expnum)
celltype_pvalue <- data.frame("celltype"=c("SSC","Diff.SPG","Early SPC","Late SPC",
                                "RS","Elongating","Elongated",
                                "Sertoli","Leydig","PMC","VSMC","EC",
                                "Macrophages","T-cell"))
for (i in c(1:14)) {
  fishertest <- fisher.test(filter(celltype_expratio,celltype==celltype_pvalue$celltype[i])[c(1,3),c(3,6)])
  celltype_pvalue[i,2] <- fishertest[["p.value"]]
  if (celltype_pvalue[i,2]>=0.05) {
    celltype_pvalue[i,3] <- "N.S."
  }
  if (celltype_pvalue[i,2]<0.05) {
    celltype_pvalue[i,3] <- "*"
  }
  if (celltype_pvalue[i,2]<0.01) {
    celltype_pvalue[i,3] <- "**"
  }
  if (celltype_pvalue[i,2]<0.001) {
    celltype_pvalue[i,3] <- "***"
  }
}
colnames(celltype_pvalue) <- c("celltype","pvalue","anno")
data.table::fwrite(celltype_pvalue,file ="/home/zhzhang/PG/scRNAseq/1expratio/Humantestis.lnc_celltype_exp_percent.diffpvalue.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#画图
ppcell <- ggplot(data = celltype_expratio,aes(x=celltype,y=expratio*100))+
  geom_col(aes(fill=type),width = 0.5,position = "dodge")+
  geom_signif(annotations=celltype_pvalue$anno,y_position=100,tip_length = 0,
              xmin = c(1:14),xmax = c(1.166:14.166))+
  scale_fill_manual(values=c("#BB5A5D","#7197AD","#FAA465","#8491B4"),
                    limits=c("Protein-coding","Pseudogene-derived lncRNA","Non-pseudogene-derived lncRNA"))+
  theme_half_open()+
  scale_y_continuous(limits = c(0,100))+
  labs(x = NULL, y ="Expression proportion (%)",fill = NULL)+
  theme(axis.title = element_text(size = 20),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),
        legend.position = "none")+
  theme(axis.text.x = element_text(angle =30))+
  theme(axis.text.x = element_text(vjust = 1,hjust=1))
ggsave("/home/zhzhang/PG/scRNAseq/1expratio/Humantestis.lnc_celltype_exp_percent.png", 
       ppcell,width = 12, height = 5,dpi=1200, units = "in", device='png',bg = "transparent")
#输出每个celltype表达的genename及其type
ctexp_genenametype <- filter(allcelltype_genepct,exp==1)%>%
  select(1,2,4)%>%
  mutate(ctagn=paste(celltype,genename,sep = "----"))
data.table::fwrite(ctexp_genenametype,file ="/home/zhzhang/PG/scRNAseq/1expratio/Humantestis.celltypeexp_genename_type.txt",sep = '\t',row.names = F,quote = F,col.names = T)


```


##### \[6.\] celltype-level具有表达的三类基因在细胞群中的阳性细胞比例对比
```r
#对比每类细胞中，表达的三类基因的阳性细胞比例
#导入基因name分类
genename_class <- read.delim("/home/zhzhang/PG/scRNAseq/genename.type.human.txt")
#导入全部基因在各celltype的阳性细胞比例，并添加基因分类
allcelltype_genepct <- read.delim("~/PG/scRNAseq/seurat/ALLsample.allcelltype.markergene.byRNA.txt")%>%
  select(6,7,3)
colnames(allcelltype_genepct) <- c("celltype","genename","pct.1")
allcelltype_genepct <- left_join(allcelltype_genepct,genename_class,by="genename")
allcelltype_genepct <- allcelltype_genepct[is.na(allcelltype_genepct$type)==F,]
#基因在某细胞群中阳性细胞比例>=0.01即为在该细胞群中表达(后续储存表达的基因列表)
allcelltype_genepct <- mutate(allcelltype_genepct,exp=0)
allcelltype_genepct$exp[allcelltype_genepct$pct.1>=0.01] <- 1
allcelltype_genepct <- filter(allcelltype_genepct,exp==1)
allcelltype_genepct$type <- factor(allcelltype_genepct$type,levels = c("Protein-coding","Pseudogene-derived lncRNA","Non-pseudogene-derived lncRNA"))
allcelltype_genepct$celltype <- factor(allcelltype_genepct$celltype,
                                     levels = c("SSC","Diff.SPG","Early SPC","Late SPC",
                                                "RS","Elongating","Elongated",
                                                "Sertoli","Leydig","PMC","VSMC","EC",
                                                "Macrophages","T-cell"))
data.table::fwrite(allcelltype_genepct,file ="/home/zhzhang/PG/scRNAseq/1expratio/Humantestis.celltypeexp_gene_poscellratio.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#统计每类细胞type中，表达的三类基因的阳性细胞比例中位数
tj <- group_by(allcelltype_genepct,celltype,type)%>%
  summarise(median=median(pct.1),mean=mean(pct.1))
data.table::fwrite(tj,file ="/home/zhzhang/PG/scRNAseq/1expratio/Humantestis.celltypeexp_gene_poscellratio.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#验证人类睾丸每个celltype中pglnc阳性细胞比例高于npglnc阳性细胞比例的显著性（wilcoxon）
celltype_pvalue <- data.frame("celltype"=c("SSC","Diff.SPG","Early SPC","Late SPC",
                                           "RS","Elongating","Elongated",
                                           "Sertoli","Leydig","PMC","VSMC","EC",
                                           "Macrophages","T-cell"))
for (i in c(1:14)) {
  test <- wilcox.test(filter(allcelltype_genepct,celltype==celltype_pvalue$celltype[i] & type=="Pseudogene-derived lncRNA")$pct.1,
                      filter(allcelltype_genepct,celltype==celltype_pvalue$celltype[i] & type=="Non-pseudogene-derived lncRNA")$pct.1)
  celltype_pvalue[i,2] <- test[["p.value"]]
  if (celltype_pvalue[i,2]>=0.05) {
    celltype_pvalue[i,3] <- "N.S."
  }
  if (celltype_pvalue[i,2]<0.05) {
    celltype_pvalue[i,3] <- "*"
  }
  if (celltype_pvalue[i,2]<0.01) {
    celltype_pvalue[i,3] <- "**"
  }
  if (celltype_pvalue[i,2]<0.001) {
    celltype_pvalue[i,3] <- "***"
  }
}
colnames(celltype_pvalue) <- c("celltype","pvalue","anno")
data.table::fwrite(celltype_pvalue,file ="/home/zhzhang/PG/scRNAseq/1expratio/Humantestis.celltypeexp_gene_poscellratio.pganpglnc_diffpvalue.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#画图
expcell <- ggplot(data = allcelltype_genepct,aes(x=celltype,y=pct.1))+
  geom_boxplot(aes(fill=type),width = 0.5,outlier.alpha = 0)+
  geom_signif(annotations=celltype_pvalue$anno,y_position=1,tip_length = 0,
              xmin = c(1:14),xmax = c(1.166:14.166))+
  scale_fill_manual(values=c("#BB5A5D","#7197AD","#FAA465","#8491B4"),
                    limits=c("Protein-coding","Pseudogene-derived lncRNA","Non-pseudogene-derived lncRNA"))+
  theme_half_open()+
  scale_y_continuous(limits = c(0,1))+
  labs(x = NULL, y ="Positive cell ratio",fill = NULL)+
  theme(axis.title = element_text(size = 20),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),
        legend.position = "none")+
  theme(axis.text.x = element_text(angle =30))+
  theme(axis.text.x = element_text(vjust = 1,hjust=1))
ggsave("/home/zhzhang/PG/scRNAseq/1expratio/Humantestis.celltypeexp_gene_poscellratio.png", 
       expcell,width = 11, height = 5,dpi=1200, units = "in", device='png',bg = "transparent")



```


##### \[7.\] celltype-level具有表达的三类基因mean expression对比
```r
ALLsample <- readRDS("/home/zhzhang/PG/scRNAseq/seurat/celltype_ALLsample.rds")
#对比每类细胞中表达的三类基因的平均表达量
#计算全部基因在所有细胞类型中的平均表达量(gene-celltype平均表达量矩阵)
ALLgene.meanexp <- AverageExpression(ALLsample,assays = "RNA",slot = "data")%>%
  data.frame()
ALLgene.meanexp_chang <- rbind(data.frame("meanexp"=ALLgene.meanexp$RNA.Elongating,"celltype"="Elongating","genename"=row.names(ALLgene.meanexp)),
                               data.frame("meanexp"=ALLgene.meanexp$RNA.Diff.SPG,"celltype"="Diff.SPG","genename"=row.names(ALLgene.meanexp)),
                               data.frame("meanexp"=ALLgene.meanexp$RNA.SSC,"celltype"="SSC","genename"=row.names(ALLgene.meanexp)),
                               data.frame("meanexp"=ALLgene.meanexp$RNA.Elongated,"celltype"="Elongated","genename"=row.names(ALLgene.meanexp)),
                               data.frame("meanexp"=ALLgene.meanexp$RNA.EC,"celltype"="EC","genename"=row.names(ALLgene.meanexp)),
                               data.frame("meanexp"=ALLgene.meanexp$RNA.PMC,"celltype"="PMC","genename"=row.names(ALLgene.meanexp)),
                               data.frame("meanexp"=ALLgene.meanexp$RNA.Early.SPC,"celltype"="Early SPC","genename"=row.names(ALLgene.meanexp)),
                               data.frame("meanexp"=ALLgene.meanexp$RNA.RS,"celltype"="RS","genename"=row.names(ALLgene.meanexp)),
                               data.frame("meanexp"=ALLgene.meanexp$RNA.Sertoli,"celltype"="Sertoli","genename"=row.names(ALLgene.meanexp)),
                               data.frame("meanexp"=ALLgene.meanexp$RNA.VSMC,"celltype"="VSMC","genename"=row.names(ALLgene.meanexp)),
                               data.frame("meanexp"=ALLgene.meanexp$RNA.Leydig,"celltype"="Leydig","genename"=row.names(ALLgene.meanexp)),
                               data.frame("meanexp"=ALLgene.meanexp$RNA.Late.SPC,"celltype"="Late SPC","genename"=row.names(ALLgene.meanexp)),
                               data.frame("meanexp"=ALLgene.meanexp$RNA.Macrophages,"celltype"="Macrophages","genename"=row.names(ALLgene.meanexp)),
                               data.frame("meanexp"=ALLgene.meanexp$RNA.T.cell,"celltype"="T-cell","genename"=row.names(ALLgene.meanexp)))
#导入每个celltype表达的genname列表
cte_gnt <- read.delim("/home/zhzhang/PG/scRNAseq/1expratio/Humantestis.celltypeexp_genename_type.txt")
#添加基因类型，去除每个细胞type中无表达基因
ALLgene.meanexp_chang <- left_join(mutate(ALLgene.meanexp_chang,ctagn=paste(celltype,genename,sep = "----")),
                                   select(cte_gnt,ctagn,type),by="ctagn")
ALLgene.meanexp_chang <- ALLgene.meanexp_chang[is.na(ALLgene.meanexp_chang$type)==F,]
ALLgene.meanexp_chang$type <- factor(ALLgene.meanexp_chang$type,levels = c("Protein-coding","Pseudogene-derived lncRNA","Non-pseudogene-derived lncRNA"))
ALLgene.meanexp_chang$celltype <- factor(ALLgene.meanexp_chang$celltype,
                                     levels = c("SSC","Diff.SPG","Early SPC","Late SPC",
                                                "RS","Elongating","Elongated",
                                                "Sertoli","Leydig","PMC","VSMC","EC",
                                                "Macrophages","T-cell"))
data.table::fwrite(ALLgene.meanexp_chang,file ="/home/zhzhang/PG/scRNAseq/1expratio/Humantestis.celltypeexp_gene_meanexp.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#统计
celltype_exp <- group_by(ALLgene.meanexp_chang,celltype,type)%>%
  summarise(mean=mean(meanexp),median=median(meanexp))
data.table::fwrite(celltype_exp,file ="/home/zhzhang/PG/scRNAseq/1expratio/Humantestis.celltypeexp_gene_meanexp.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#验证人类睾丸每个celltype中pglnc平均表达量高于npglnc平均表达量的显著性（wilcox）
celltype_pvalue <- data.frame("celltype"=c("SSC","Diff.SPG","Early SPC","Late SPC",
                                           "RS","Elongating","Elongated",
                                           "Sertoli","Leydig","PMC","VSMC","EC",
                                           "Macrophages","T-cell"))
for (i in c(1:14)) {
  test <- wilcox.test(filter(ALLgene.meanexp_chang,celltype==celltype_pvalue$celltype[i] & type=="Pseudogene-derived lncRNA")$meanexp,
                      filter(ALLgene.meanexp_chang,celltype==celltype_pvalue$celltype[i] & type=="Non-pseudogene-derived lncRNA")$meanexp)
  celltype_pvalue[i,2] <- test[["p.value"]]
  if (celltype_pvalue[i,2]>=0.05) {
    celltype_pvalue[i,3] <- "N.S."
  }
  if (celltype_pvalue[i,2]<0.05) {
    celltype_pvalue[i,3] <- "*"
  }
  if (celltype_pvalue[i,2]<0.01) {
    celltype_pvalue[i,3] <- "**"
  }
  if (celltype_pvalue[i,2]<0.001) {
    celltype_pvalue[i,3] <- "***"
  }
}
colnames(celltype_pvalue) <- c("celltype","pvalue","anno")
data.table::fwrite(celltype_pvalue,file ="/home/zhzhang/PG/scRNAseq/1expratio/Humantestis.celltypeexp_gene_meanexp.pganpglnc_diffpvalue.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#画图
expcell <- ggplot(data = ALLgene.meanexp_chang,aes(x=celltype,y=log1p(meanexp)))+
  geom_boxplot(aes(fill=type),width = 0.5,outlier.alpha = 0)+
  geom_signif(annotations=celltype_pvalue$anno,y_position=1.2,tip_length = 0,
              xmin = c(1:14),xmax = c(1.166:14.166))+
  scale_fill_manual(values=c("#BB5A5D","#7197AD","#FAA465","#8491B4"),
                    limits=c("Protein-coding","Pseudogene-derived lncRNA","Non-pseudogene-derived lncRNA"))+
  theme_half_open()+
  coord_cartesian(ylim = c(0,1.25))+
  scale_y_continuous(breaks = c(0,0.25,0.5,0.75,1.0,1.25))+
  labs(x = NULL, y ="Average Expression",fill = NULL)+
  theme(axis.title = element_text(size = 20),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),
        legend.position = "none")+
  theme(axis.text.x = element_text(angle =30))+
  theme(axis.text.x = element_text(vjust = 1,hjust=1))
ggsave("/home/zhzhang/PG/scRNAseq/1expratio/Humantestis.celltypeexp_gene_meanexp.png", 
       expcell,width = 11, height = 5,dpi=1200, units = "in", device='png',bg = "transparent")


```


##### \[12.\]celltype表达的三类基因marker gene占比
```r
#celltype表达的三类基因marker gene占比
#导入每个celltype表达的genname列表
cte_gnt <- read.delim("/home/zhzhang/PG/scRNAseq/1expratio/Humantestis.celltypeexp_genename_type.txt")
#导入celltype的markergene
allcelltype_marker <- read.delim("~/PG/scRNAseq/seurat/ALLsample.allcelltype.markergene.byRNA.txt")%>%
  mutate(marker=0)
#Log2FC > 1 and padj < 0.05 and pct1>0.25为marker gene
allcelltype_marker$marker[allcelltype_marker$p_val_adj<0.05 & allcelltype_marker$avg_log2FC>=1 & allcelltype_marker$pct.1>=0.25] <- 1
#添加基因类型
allcelltype_marker <- left_join(mutate(allcelltype_marker,ctagn=base::paste(cluster,gene,sep = "----")),
                                select(cte_gnt,3,4),by="ctagn")%>%
  select(-ctagn)
allcelltype_marker <- allcelltype_marker[is.na(allcelltype_marker$type)==F,]
#统计每个celltype表达的三类基因以及作为markergene的三类基因
tj <- group_by(allcelltype_marker,cluster,type)%>%
  summarise(num=n(),markernum=sum(marker))%>%
  mutate(nomarker=num-markernum,markerratio=markernum/num)
colnames(tj)[1] <- "celltype"
tj$type <- factor(tj$type,levels = c("Protein-coding","Pseudogene-derived lncRNA","Non-pseudogene-derived lncRNA"))
tj$celltype <- factor(tj$celltype,
                      levels = c("SSC","Diff.SPG","Early SPC","Late SPC",
                                 "RS","Elongating","Elongated",
                                 "Sertoli","Leydig","PMC","VSMC","EC",
                                 "Macrophages","T-cell"))
data.table::fwrite(tj,file ="/home/zhzhang/PG/scRNAseq/1expratio/Humantestis.celltypeexp_gene_markerratio.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#验证人类睾丸每个celltype中pglnc表达比例高于npglnc表达比例的显著性（fishertest）
celltype_pvalue <- data.frame("celltype"=c("SSC","Diff.SPG","Early SPC","Late SPC",
                                           "RS","Elongating","Elongated",
                                           "Sertoli","Leydig","PMC","VSMC","EC",
                                           "Macrophages","T-cell"))
for (i in c(1:14)) {
  fishertest <- fisher.test(filter(tj,celltype==celltype_pvalue$celltype[i])[c(1,3),c(4,5)])
  celltype_pvalue[i,2] <- fishertest[["p.value"]]
  if (celltype_pvalue[i,2]>=0.05) {
    celltype_pvalue[i,3] <- "N.S."
  }
  if (celltype_pvalue[i,2]<0.05) {
    celltype_pvalue[i,3] <- "*"
  }
  if (celltype_pvalue[i,2]<0.01) {
    celltype_pvalue[i,3] <- "**"
  }
  if (celltype_pvalue[i,2]<0.001) {
    celltype_pvalue[i,3] <- "***"
  }
}
colnames(celltype_pvalue) <- c("celltype","pvalue","anno")
data.table::fwrite(celltype_pvalue,file ="/home/zhzhang/PG/scRNAseq/1expratio/Humantestis.celltypeexp_gene_markerratio.diffpvalue.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#画图
pp <- ggplot(data = tj,aes(x=celltype,y=markerratio*100))+
  geom_col(aes(fill=type),width = 0.5,position = "dodge")+
  geom_signif(annotations=celltype_pvalue$anno,y_position=15,tip_length = 0,
              xmin = c(1:14),xmax = c(1.166:14.166))+
  scale_fill_manual(values=c("#BB5A5D","#7197AD","#FAA465","#8491B4"),
                    limits=c("Protein-coding","Pseudogene-derived lncRNA","Non-pseudogene-derived lncRNA"))+
  theme_half_open()+
  scale_y_continuous(limits = c(0,15))+
  labs(x = NULL, y ="Marker gene proportion (%)",fill = NULL)+
  theme(axis.title = element_text(size = 20),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),
        legend.position = "none")+
  theme(axis.text.x = element_text(angle =30))+
  theme(axis.text.x = element_text(vjust = 1,hjust=1))
ggsave("/home/zhzhang/PG/scRNAseq/1expratio/Humantestis.celltypeexp_gene_markerratio.png", 
       pp,width = 12, height = 5,dpi=1200, units = "in", device='png',bg = "transparent")
#保存各celltype markergene
mg <- filter(allcelltype_marker,marker==1)
data.table::fwrite(mg,file ="/home/zhzhang/PG/scRNAseq/1expratio/Humantestis.allcelltype.markergene.txt",sep = '\t',row.names = F,quote = F,col.names = T)


```




##### \[8.\] 在至少一个celltype表达的三类基因的细胞特异性（在几种celltype中表达）
```r
#对比检测到表达的三类基因的细胞特异性（在几种celltype中表达）
#导入每个celltype表达的genname列表
cte_gnt <- read.delim("/home/zhzhang/PG/scRNAseq/1expratio/Humantestis.celltypeexp_genename_type.txt")
#统计每个基因在几个细胞type中表达
gene_celltypenum <- group_by(mutate(cte_gnt,num=1),genename,type)%>%
  summarise(celltypenum=sum(num))
gene_celltypenum$type <- factor(gene_celltypenum$type,levels = rev(c("Protein-coding","Pseudogene-derived lncRNA","Non-pseudogene-derived lncRNA")))
data.table::fwrite(gene_celltypenum,file ="/home/zhzhang/PG/scRNAseq/2cellspecify/Humantestis.celltypeexp_gene_celltypenum.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#统计
tj <- group_by(gene_celltypenum,type)%>%
  summarise(median=median(celltypenum),mean=mean(celltypenum))
data.table::fwrite(tj,file ="/home/zhzhang/PG/scRNAseq/2cellspecify/Humantestis.celltypeexp_gene_celltypenum.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#plot
library(gghalves)
wp1 <- signif(wilcox.test(filter(gene_celltypenum,type=="Pseudogene-derived lncRNA")$celltypenum,
                         filter(gene_celltypenum,type=="Non-pseudogene-derived lncRNA")$celltypenum)[["p.value"]],
             2)
wp2 <- signif(wilcox.test(filter(gene_celltypenum,type=="Pseudogene-derived lncRNA")$celltypenum,
                   filter(gene_celltypenum,type=="Protein-coding")$celltypenum)[["p.value"]],
             2)
cellnumden <- ggplot(data = gene_celltypenum,aes(x=type,y=celltypenum))+
  geom_half_violin(width=1,side = "r",position = position_nudge(x=0.2),aes(color=type,fill=type))+
  geom_point(data = filter(gene_celltypenum,type=="Protein-coding"),stroke=0,
             alpha=0.8/(18264/952/5),size=0.5,position = position_jitter(width=0.2),aes(color=type))+
  geom_point(data = filter(gene_celltypenum,type=="Non-pseudogene-derived lncRNA"),stroke=0,
             alpha=0.8/(25582/952/5),size=0.5,position = position_jitter(width=0.2),aes(color=type))+
  geom_point(data = filter(gene_celltypenum,type=="Pseudogene-derived lncRNA"),stroke=0,
             alpha=0.8,size=0.5,position = position_jitter(width=0.2),aes(color=type))+
  geom_boxplot(fatten=3,width=0.05,position = position_nudge(x=0.2),fill="white",outlier.alpha = 0)+
  coord_flip()+
  scale_x_discrete(labels = c("Non-pseudogene-\nderived lncRNA","Pseudogene-\nderived lncRNA","Protein-coding"))+
  scale_fill_manual(values=c("#BB5A5D","#7197AD","#FAA465","#8491B4"),
                    limits=c("Protein-coding","Pseudogene-derived lncRNA","Non-pseudogene-derived lncRNA"))+
  scale_color_manual(values=c("#BB5A5D","#7197AD","#FAA465","#8491B4"),
                    limits=c("Protein-coding","Pseudogene-derived lncRNA","Non-pseudogene-derived lncRNA"))+
  theme_half_open()+
  labs(x =NULL, y ="Number of cell types",fill = NULL,color=NULL)+
  theme(axis.title = element_text(size = 20),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),legend.position = "none")+
  geom_signif(annotations=c(wp1,wp2),y_position=15,tip_length = 0,
              xmin = c(1.2,2.2),xmax = c(1.8,2.8))
ggsave("/home/zhzhang/PG/scRNAseq/2cellspecify/Humantestis.celltypeexp_gene_celltypenum.png", 
       cellnumden,width = 7, height = 5,dpi=1200, units = "in", device='png',bg = "transparent")



```








##### \[11.\]germ cell拟时分析
```r
###细胞轨迹的完全建立
ALLsample <- readRDS("/home/zhzhang/PG/scRNAseq/seurat/celltype_ALLsample.rds")
ALLsample@active.assay <- "RNA"
ALLsample <- subset(ALLsample,cell_type=="SSC"|cell_type=="Diff.SPG"|cell_type=="Early SPC"|cell_type=="Late SPC"|cell_type=="RS"|cell_type=="Elongating"|cell_type=="Elongated")
gc()
#提取表达矩阵
data <- LayerData(ALLsample, assay = 'RNA', layer = 'counts')
#提取细胞信息
cell_metadata <- ALLsample@meta.data
#提取基因信息（genename）甚至可以包括gene类型、gc含量等
gene_annotation <- data.frame(gene_short_name = row.names(ALLsample),row.names = row.names(ALLsample))
#构建CDS对象
cds <- new_cell_data_set(data,
                         cell_metadata = cell_metadata,
                         gene_metadata = gene_annotation)
#preprocess_cds函数相当于seurat中NormalizeData+ScaleData+RunPCA
cds <- preprocess_cds(cds, num_dim = 15)
#使用以下命令查看每台PC解释的变异分数（elbow plot）
plot_pc_variance_explained(cds)
#UMAP降维
cds <- reduce_dimension(cds,reduction_method="UMAP",preprocess_method = "PCA")
#聚类
cds <- cluster_cells(cds,reduction_method="UMAP")
#轨迹建立
cds <- learn_graph(cds,use_partition=F)
#确定轨迹起点
get_earliest_principal_node <- function(cds, celltype="SSC"){
  cell_ids <- which(colData(cds)[, "cell_type"] == celltype)
  
  closest_vertex <-
    cds@principal_graph_aux[["UMAP"]]$pr_graph_cell_proj_closest_vertex
  closest_vertex <- as.matrix(closest_vertex[colnames(cds), ])
  root_pr_nodes <-
    igraph::V(principal_graph(cds)[["UMAP"]])$name[as.numeric(names
                                                              (which.max(table(closest_vertex[cell_ids,]))))]
  
  root_pr_nodes
}
cds <- order_cells(cds, root_pr_nodes=get_earliest_principal_node(cds))
save_monocle_objects(cds,directory_path = "/home/zhzhang/PG/scRNAseq/monocle3/cdsceshi.cds")
#plot
cds@colData@listData[["cell_type"]] <- factor(cds@colData@listData[["cell_type"]],
                                              levels =c("SSC","Diff.SPG","Early SPC",
                                                        "Late SPC","RS","Elongating","Elongated"))
track_celltype <- plot_cells(cds,
                             color_cells_by = "cell_type",
                             label_groups_by_cluster=FALSE,
                             label_leaves=F,
                             label_branch_points=F,
                             label_cell_groups = F,
                             trajectory_graph_segment_size=1)+
  scale_color_manual(values = c("#2A6434","#50A15C","#7DBB78",
                                "#ACD18E","#A6C9DD","#619BC0",
                                "#337CA5"))+
  cowplot::theme_half_open()+
  theme(axis.title = element_text(size = 15),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 15))+
  labs(title = NULL,x="UMAP-1",y="UMAP-2")+
  theme(legend.text = element_text(size = 14))
ggsave("/home/zhzhang/PG/scRNAseq/monocle3/Humantestis.germcelltype_track_umap.pdf", 
       track_celltype,width = 8, height = 6,dpi=1200, units = "in", device='pdf',bg = "transparent")
pt_celltype <- plot_cells(cds, color_cells_by = "pseudotime", 
                          label_leaves = FALSE,  label_branch_points = FALSE,trajectory_graph_segment_size=1)+
  cowplot::theme_half_open()+
  theme(axis.title = element_text(size = 15),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 15))+
  labs(title = NULL,x="UMAP-1",y="UMAP-2")+
  theme(legend.text = element_text(size = 15))+
  theme(axis.title = element_text(size = 15),axis.text.y  = element_text(size = 0),
        axis.text.x = element_text(size = 0))+
  theme(axis.ticks = element_line(colour = NA))
ggsave("/home/zhzhang/PG/scRNAseq/monocle3/Humantestis.germcelltype_pseudotime_umap.pdf", 
       pt_celltype,width = 8, height = 6,dpi=1200, units = "in", device='pdf',bg = "transparent")

#pseudotime地毯图
cell_pt <- data.frame(pseudotime=cds@principal_graph_aux@listData[["UMAP"]][["pseudotime"]],
                      cell_type=cds@colData@listData[["cell_type"]])%>%
  tibble::rownames_to_column("barcode")
ggplot(data=cell_pt,aes(x=pseudotime))+
  geom_rug(aes(color=cell_type))



```
```r
###确定在细胞轨迹上差异表达的基因
#确定在至少一种germ cell表达的,在细胞轨迹上差异表达的基因
cds <- load_monocle_objects(directory_path='/home/zhzhang/PG/scRNAseq/monocle3/cdsceshi.cds')
cds@colData@listData[["cell_type"]] <- factor(cds@colData@listData[["cell_type"]],
                                              levels =c("SSC","Diff.SPG","Early SPC",
                                                        "Late SPC","RS","Elongating","Elongated"))
#graph_test分析
Track_genes <- graph_test(cds, neighbor_graph="principal_graph", cores=2)
data.table::fwrite(Track_genes,file ="/home/zhzhang/PG/scRNAseq/monocle3/Humantestis.germcelltype_pseudotimeDEG.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#导入celltype表达的基因
celltypeexp_genename_type <- read.delim("~/PG/scRNAseq/1expratio/Humantestis.celltypeexp_genename_type.txt")
#筛选出在至少一种germ cell表达的基因
germexp_gene <- filter(celltypeexp_genename_type,celltype=="SSC"|celltype=="Diff.SPG"|celltype=="Early SPC"|celltype=="Late SPC"|celltype=="RS"|celltype=="Elongating"|celltype=="Elongated")%>%
  distinct(genename,type)
#基因莫兰分析结果添加所属类型，不属于三类基因的基因删除
gmoran <- left_join(mutate(Track_genes,genename=gene_short_name),germexp_gene,by="genename")
gmoran <- gmoran[is.na(gmoran$type)==F,]
gmoran <- select(gmoran,-1,-5)
data.table::fwrite(gmoran,file ="/home/zhzhang/PG/scRNAseq/monocle3/Humantestis.germcellexp_gene_moran.txt",sep = '\t',row.names = F,quote = F,col.names = T)
gmoran$type <- factor(gmoran$type,levels = c("Protein-coding","Pseudogene-derived lncRNA","Non-pseudogene-derived lncRNA"))
#统计三类基因莫兰指数
tj <- group_by(gmoran,type)%>%
  summarise(median=median(morans_I),mean=mean(morans_I))
data.table::fwrite(tj,file ="/home/zhzhang/PG/scRNAseq/monocle3/Humantestis.germcellexp_gene_moranI.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#plot对比三类基因莫兰指数（代表：表达量在细胞轨迹上的空间自相关性）
phs <- ggplot(data=gmoran, aes(x=type,y=morans_I))+
  geom_violin(width=1,alpha=0.9,aes(fill=type,color=type))+
  geom_boxplot(fatten = 3,outlier.alpha = 0,width=0.1,notch=T)+
  geom_signif(map_signif_level=T,y_position=c(0.8,0.75),tip_length = 0.01,
              comparisons = list(c("Protein-coding","Pseudogene-derived lncRNA"),
                                 c("Non-pseudogene-derived lncRNA","Pseudogene-derived lncRNA")
              ))+
  scale_fill_manual(values=c("#BB5A5D","#7197AD","#FAA465","#8491B4"),
                    limits=c("Protein-coding","Pseudogene-derived lncRNA",
                             "Non-pseudogene-derived lncRNA"))+
  scale_color_manual(values=c("#BB5A5D","#7197AD","#FAA465","#8491B4"),
                     limits=c("Protein-coding","Pseudogene-derived lncRNA",
                              "Non-pseudogene-derived lncRNA"))+
  theme_half_open()+
  scale_y_continuous(breaks = c(0,0.2,0.4,0.6,0.8))+
  scale_x_discrete(labels = c("Protein-coding","Pseudogene-\nderived lncRNA",
                              "Non-pseudogene-\nderived lncRNA"))+
  labs(y = "Moran's I",x =NULL,fill = NULL,color = NULL)+
  theme(axis.title = element_text(size = 20),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 0),legend.position = "none") +
  theme(axis.text.x = element_text(angle =30)) +
  theme(axis.text.x = element_text(hjust=1,vjust = 1)) +
  theme(axis.ticks.x = element_line(colour = NA))
ggsave("/home/zhzhang/PG/scRNAseq/monocle3/Humantestis.germcellexp_gene_moranI.pdf",
       phs,width = 4.8, height = 4,dpi=1200, units = "in", device='pdf',bg = "transparent")
#筛选在细胞轨迹上差异表达的基因（在细胞轨迹上表达量具有空间自相关性的基因）（q值<0.001且莫兰指数>0.2）
gmoran <- mutate(gmoran,ptde=0)
gmoran$ptde[gmoran$q_value < 1e-3 & gmoran$morans_I > 0.2] <- 1
data.table::fwrite(gmoran,file ="/home/zhzhang/PG/scRNAseq/monocle3/Humantestis.germcellexp_gene_moran.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#统计在细胞轨迹上DE基因比例
tjptratio <- group_by(gmoran,type)%>%
  summarise(num=n(),ptdenum=sum(ptde))%>%
  mutate(ptderatio=ptdenum/num,noptdenum=num-ptdenum)
data.table::fwrite(tjptratio,file ="/home/zhzhang/PG/scRNAseq/monocle3/Humantestis.germcellexp_gene_ptDEratio.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#（验证pglnc相比于npglnc基因在细胞轨迹上DE比例更高的显著性）
#（pglnc-pc基因）p-value < 2.2e-16
fisher.test(tjptratio[c(1,2),c(3,5)])
#（pglnc-npglnc基因）p-value < 2.2e-16
fisher.test(tjptratio[c(2,3),c(3,5)])
#plot时序DE基因比例差异
cedif <- ggplot(data = tjptratio,aes(x=type,y=ptderatio*100))+
  geom_col(aes(fill=type),width = 0.5)+
  geom_signif(annotations=c("***","***"),y_position=c(56,35),tip_length = 0,
              xmin = c(1,2),xmax = c(2,3))+
  scale_fill_manual(values=c("#BB5A5D","#7197AD","#FAA465","#8491B4"),
                    limits=c("Protein-coding","Pseudogene-derived lncRNA",
                             "Non-pseudogene-derived lncRNA"))+
  scale_x_discrete(labels = c("Protein-coding","Pseudogene-\nderived lncRNA",
                              "Non-pseudogene-\nderived lncRNA"))+
  scale_y_continuous(breaks = c(0,10,20,30,40,50))+
  theme_half_open()+
  labs(x = NULL, y ="The proportion of DE genes\nalong the single-cell trajectory (%)",fill = NULL)+
  theme(axis.title = element_text(size = 17),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 0),
        legend.position = "none")+
  theme(axis.text.x = element_text(angle =30)) +
  theme(axis.text.x = element_text(hjust=1,vjust = 1))+
  theme(axis.ticks.x = element_line(colour = NA))
ggsave("/home/zhzhang/PG/scRNAseq/monocle3/Humantestis.germcellexp_gene_ptDEratio.pdf", 
       cedif,width = 4.8, height = 4,dpi=1200, units = "in", device='pdf',bg = "transparent")


```
```r
###在细胞轨迹上差异表达的基因聚类，并与celltype关联
#导入基因moran检验结果
gmoran <- read.delim("~/PG/scRNAseq/monocle3/Humantestis.germcellexp_gene_moran.txt")
#在细胞轨迹上DEgene聚类为模块
DEtrackgene <- filter(gmoran,morans_I > 0.2 & q_value < 1e-3)$genename
gene_module_df <- find_gene_modules(cds[DEtrackgene,],cores=8,
                                    random_seed=27)
#模块与celltype关联,计算module-celltype整合表达量矩阵
cell_group_df <- tibble::tibble(cell=row.names(colData(cds)), 
                                cell_group=colData(cds)$cell_type)
agg_mat <- aggregate_gene_expression(cds, gene_module_df, cell_group_df,
                                     norm_method="log",max_agg_value=3,min_agg_value=-3,
                                     scale_agg_values =F)%>%
  as.data.frame()
row.names(agg_mat) <- base::paste("Module", row.names(agg_mat),sep = "")
agg_mat <- agg_mat[c(1,3,4,2),]
data.table::fwrite(agg_mat,file ="/home/zhzhang/PG/scRNAseq/monocle3/Humantestis.germcellexp_gene_trackDEgenemodule_celltype_exp.matrix.txt",sep = '\t',row.names = T,quote = F,col.names = T)
#折线图对比四类module整合表达量在celltype中的变化
agg_mat_forplot <- data.frame(t(agg_mat))%>%
  mutate(Module1=Module1*100/max(Module1),Module2=Module2*100/max(Module2),
         Module3=Module3*100/max(Module3),Module4=Module4*100/max(Module4))%>%
  rownames_to_column("celltype")%>%
  gather("Module","exppercent",2:5)
agg_mat_forplot$celltype <- factor(agg_mat_forplot$celltype,
                                   levels =c("SSC","Diff.SPG","Early SPC",
                                             "Late SPC","RS","Elongating","Elongated"))
xspline <- ggplot(data =agg_mat_forplot,aes(x=celltype,y=exppercent))+
  geom_point(aes(color=Module))+
  ggalt::geom_xspline((aes(color=Module,group=Module)))+
  ggsci::scale_color_npg()+
  theme_half_open()+
  labs(x = NULL, y ="Percentage of maximum\naggregation expression (%)",color = NULL)+
  theme(axis.title = element_text(size = 17),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 13))+
  theme(axis.text.x = element_text(angle =30)) +
  theme(axis.text.x = element_text(hjust=1,vjust = 1))
ggsave("/home/zhzhang/PG/scRNAseq/monocle3/Humantestis.germcellexp_gene_ptDEgene_Module_celltype_percentmaxexpline.png", 
       xspline,width = 6.5, height = 4,dpi=1200, units = "in", device='png',bg = "transparent")
dev.off()
#在细胞轨迹上DE的gene module在每种celltype中的整合表达量热图(列scale,看celltype中最大表达量的module)
ph <- pheatmap::pheatmap(agg_mat,
                   scale="row",cluster_cols = F, cluster_rows=T,
                   color=colorRampPalette(c('#4DBBD5','white','#E64B35'))(100),
                   legend_breaks=c(-1.5,-1,-0.5,0,0.5,1,1.5),
                   border_color=NA,angle_col=45,
                   fontsize_col=12,fontsize_row=12,clustering_method = "ward.D2")
ggsave("/home/zhzhang/PG/scRNAseq/monocle3/Humantestis.germcellexp_gene_ptDEgene_Module_celltype_expheatmap.png", 
       ph,width = 6, height = 3.5,dpi=1200, units = "in", device='png',bg = "transparent")
#在细胞轨迹上DE的pglnc gene module在每种celltype中的整合表达量热图(列scale,看celltype中最大表达量的module)
cds <- load_monocle_objects(directory_path='/home/zhzhang/PG/scRNAseq/monocle3/cdsceshi.cds')
cds@colData@listData[["cell_type"]] <- factor(cds@colData@listData[["cell_type"]],
                                              levels =c("SSC","Diff.SPG","Early SPC",
                                                        "Late SPC","RS","Elongating","Elongated"))
gene_module_df <- read.delim("~/PG/scRNAseq/monocle3/Humantestis.germcellexp_gene_trackDEgenemodule.txt")
cell_group_df <- tibble::tibble(cell=row.names(colData(cds)), 
                                cell_group=colData(cds)$cell_type)
gene_module_df <- filter(gene_module_df,type=="Pseudogene-derived lncRNA")
agg_mat <- aggregate_gene_expression(cds, gene_module_df, cell_group_df,
                                     norm_method="log",max_agg_value=3,min_agg_value=-3,
                                     scale_agg_values =F)%>%
  as.data.frame()
row.names(agg_mat) <- base::paste("Module", row.names(agg_mat),sep = "")
agg_mat <- agg_mat[c(2,1,4,3),]

ph <- pheatmap::pheatmap(agg_mat,
                         scale="row",cluster_cols = F, cluster_rows=T,
                         color=colorRampPalette(c('#4DBBD5','white','#E64B35'))(100),
                         legend_breaks=c(-1.5,-1,-0.5,0,0.5,1,1.5),
                         border_color=NA,angle_col=45,
                         fontsize_col=12,fontsize_row=12,clustering_method = "ward.D2")
ggsave("/home/zhzhang/PG/scRNAseq/monocle3/Humantestis.germcellexp_gene_ptDEpglnc_Module_celltype_expheatmap.pdf", 
       ph,width = 6, height = 3.5,dpi=1200, units = "in", device='pdf',bg = "transparent")
#在细胞轨迹上DE的gene module在每个cell中的整合表达量UMAP图
umap <- plot_cells(cds,
           genes=gene_module_df %>% filter(module %in% c(1,2,3,4)),
           label_cell_groups=FALSE,
           show_trajectory_graph=FALSE,group_label_size = 10)+
  theme(axis.title = element_text(size = 13),axis.text.y  = element_text(size = 13),
        axis.text.x = element_text(size = 13))
ggsave("/home/zhzhang/PG/scRNAseq/monocle3/Humantestis.germcellexp_gene_ptDEgene_Module_cell_expUMAP.png", 
       umap,width = 7, height = 6,dpi=1200, units = "in", device='png',bg = "transparent")
#统计在细胞轨迹上DE的三类基因的module分布
colnames(gene_module_df)[1] <- "genename"
gene_module_df <- left_join(gene_module_df,germexp_gene,by="genename")
data.table::fwrite(gene_module_df,file ="/home/zhzhang/PG/scRNAseq/monocle3/Humantestis.germcellexp_gene_trackDEgenemodule.txt",sep = '\t',row.names = F,quote = F,col.names = T)
type_module_table <- table(gene_module_df$type,gene_module_df$module)%>%
  data.frame()
colnames(type_module_table) <- c("type","module","num")
type_module_table <- left_join(type_module_table,select(tjptratio,1,3),by="type")%>%
  mutate(ratio=num/ptdenum)
type_module_table$type <- factor(type_module_table$type,
                                 levels = c("Protein-coding","Pseudogene-derived lncRNA","Non-pseudogene-derived lncRNA"))
data.table::fwrite(type_module_table,file ="/home/zhzhang/PG/scRNAseq/monocle3/Humantestis.germcellexp_gene_ptDEgene_3class_Moduledistribution.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#plot比例堆积柱状图，对比在细胞轨迹上DE的三类基因的module分布情况
type_module_table <- mutate(type_module_table,module=paste("Module",module,sep = ""))
md <- ggplot(data=type_module_table,aes(x=type,y=num))+
  geom_col(aes(fill=module),position = "fill",width = 0.6)+
  ggsci::scale_fill_npg()+
  scale_x_discrete(labels = c("Protein-coding","Pseudogene-\nderived lncRNA",
                              "Non-pseudogene-\nderived lncRNA"))+
  theme_half_open()+
  labs(x = NULL, y ="Module distribution",fill = NULL)+
  theme(axis.title = element_text(size = 20),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14))+
  theme(axis.text.x = element_text(angle =40)) +
  theme(axis.text.x = element_text(hjust=1,vjust = 1))+
  theme(legend.text = element_text(size = 15))
ggsave("/home/zhzhang/PG/scRNAseq/monocle3/Humantestis.germcellexp_gene_ptDEgene_3class_Moduledistribution.png", 
       md,width = 5, height = 5.5,dpi=1200, units = "in", device='png',bg = "transparent")
#fishertest检验npglnc中module2比例显著高于pglnc（p-value < 2.2e-16）
fisher.test(data.frame(x=c(84,2345),y=c(210,1574)))
#fishertest检验pglnc中module1比例显著高于npglnc（p-value < 2.2e-16）
fisher.test(data.frame(x=c(174,1030),y=c(120,2889)))

```
##### 沿细胞轨迹DEpglnc的coexp基因富集分析
```r
#导入精子发生细胞轨迹中DE的基因list
trackDEgenemodule <- read.delim("~/PG/scRNAseq/monocle3/Humantestis.germcellexp_gene_trackDEgenemodule.txt")
#导入coexp的pcgene
pglnc_addtarget <- read.delim("/home/zhzhang/PG/RNAseq/Homo_sapiens/addtarget_new/Homo_sapiens.spexp.pglnc_pcgene.coexppair.top20.txt")%>%
  select(1,2)
#导入name id对照
ID_name <- read.delim("~/PG/scRNAseq/gene.ID.name.human.txt", header=FALSE)
colnames(ID_name) <- c("lncgid","genename")
ID_name <- mutate(ID_name,genename=stringr::str_replace_all(genename,"_","-"))
#筛选出每个module中的pglnc，并转为ensembleid
pglnc_m1 <- filter(trackDEgenemodule,type=="Pseudogene-derived lncRNA")%>%
  filter(module=="1")%>%
  select(1)%>%
  left_join(ID_name,by="genename")%>%
  select(2)
pglnc_m3 <- filter(trackDEgenemodule,type=="Pseudogene-derived lncRNA")%>%
  filter(module=="3")%>%
  select(1)%>%
  left_join(ID_name,by="genename")%>%
  select(2)
pglnc_m2 <- filter(trackDEgenemodule,type=="Pseudogene-derived lncRNA")%>%
  filter(module=="2")%>%
  select(1)%>%
  left_join(ID_name,by="genename")%>%
  select(2)
pglnc_m4 <- filter(trackDEgenemodule,type=="Pseudogene-derived lncRNA")%>%
  filter(module=="4")%>%
  select(1)%>%
  left_join(ID_name,by="genename")%>%
  select(2)
#确定每个module pglnc的靶基因
pglnc_s1_tar <- left_join(pglnc_m1,pglnc_addtarget,by="lncgid")%>%
  distinct(targetid)
pglnc_s2_tar <- left_join(pglnc_m2,pglnc_addtarget,by="lncgid")%>%
  distinct(targetid)
pglnc_s3_tar <- left_join(pglnc_m3,pglnc_addtarget,by="lncgid")%>%
  distinct(targetid)
pglnc_s4_tar <- left_join(pglnc_m4,pglnc_addtarget,by="lncgid")%>%
  distinct(targetid)
pglnc_s1_tar <- data.frame("geneid"=pglnc_s1_tar[is.na(pglnc_s1_tar$targetid)==F,])
pglnc_s2_tar <- data.frame("geneid"=pglnc_s2_tar[is.na(pglnc_s2_tar$targetid)==F,])
pglnc_s3_tar <- data.frame("geneid"=pglnc_s3_tar[is.na(pglnc_s3_tar$targetid)==F,])
pglnc_s4_tar <- data.frame("geneid"=pglnc_s4_tar[is.na(pglnc_s4_tar$targetid)==F,])


b <- 'org.Hs.eg.db'
bb <- "hsa"
#GO富集分析
pglnc_s1_tar_GO <- clusterProfiler::enrichGO(gene = pglnc_s1_tar$geneid, OrgDb = b, keyType = "ENSEMBL", ont = 'BP', pAdjustMethod = 'fdr', pvalueCutoff = 1, qvalueCutoff = 1)
pglnc_s1_tar_GO_result <- pglnc_s1_tar_GO@result%>%
  filter(p.adjust<0.05)

pglnc_s3_tar_GO <- clusterProfiler::enrichGO(gene = pglnc_s3_tar$geneid, OrgDb = b, keyType = "ENSEMBL", ont = 'BP', pAdjustMethod = 'fdr', pvalueCutoff = 1, qvalueCutoff = 1)
pglnc_s3_tar_GO_result <- pglnc_s3_tar_GO@result%>%
  filter(p.adjust<0.05)

pglnc_s2_tar_GO <- clusterProfiler::enrichGO(gene = pglnc_s2_tar$geneid, OrgDb = b, keyType = "ENSEMBL", ont = 'BP', pAdjustMethod = 'fdr', pvalueCutoff = 1, qvalueCutoff = 1)
pglnc_s2_tar_GO_result <- pglnc_s2_tar_GO@result%>%
  filter(p.adjust<0.05)

pglnc_s4_tar_GO <- clusterProfiler::enrichGO(gene = pglnc_s4_tar$geneid, OrgDb = b, keyType = "ENSEMBL", ont = 'BP', pAdjustMethod = 'fdr', pvalueCutoff = 1, qvalueCutoff = 1)
pglnc_s4_tar_GO_result <- pglnc_s4_tar_GO@result%>%
  filter(p.adjust<0.05)
#合并储存
enrich <- rbind(pglnc_s1_tar_GO_result,pglnc_s3_tar_GO_result,pglnc_s2_tar_GO_result,pglnc_s4_tar_GO_result)%>%
  mutate(Description=Hmisc::capitalize(Description))%>%
  mutate(Description=do::Replace(data=Description,pattern=c("NcRNA:ncRNA","MRNA:mRNA","RRNA:rRNA","PiRNA:piRNA")))
data.table::fwrite(enrich,file ="/home/zhzhang/PG/scRNAseq/monocle3/Humantestis.germcellexp_trackDE_modulepglnc_top20coexp.enrich.txt",
                   sep = '\t',row.names = F,quote = F,col.names = T)
#提取基因数top10的通路画图展示
filter1_enrich <- top_n(pglnc_s1_tar_GO_result,10,Count)%>%
  mutate(type="Module1")
filter2_enrich <- top_n(pglnc_s2_tar_GO_result,12,Count)%>%
  mutate(type="Module2")
filter3_enrich <- top_n(pglnc_s3_tar_GO_result,10,Count)%>%
  mutate(type="Module3")
filter4_enrich <- top_n(pglnc_s4_tar_GO_result,10,Count)%>%
  mutate(type="Module4")

parent_enrich <- rbind(filter1_enrich,filter3_enrich,filter2_enrich,filter4_enrich)%>%
  mutate(Description=Hmisc::capitalize(Description))%>%
  mutate(Description=do::Replace(data=Description,pattern=c("NcRNA:ncRNA","MRNA:mRNA","RRNA:rRNA","PiRNA:piRNA")))%>%
  filter(Description!="Cilium or flagellum-dependent cell motility")%>%
  filter(Description!="Cellular process involved in reproduction in multicellular organism")%>%
  group_by(type)%>%
  dplyr::top_n(10,-log10(p.adjust))

parent_enrich$type <- factor(parent_enrich$type,
                             levels = c("Module1","Module3","Module2","Module4"))
parent_enrich$Description <- factor(parent_enrich$Description,
                                    levels = distinct(data.frame("Description"=parent_enrich$Description),Description)$Description)
#plot
enrich <- ggplot(data = parent_enrich,aes(x=type,y=Description))+
  geom_point(aes(color=-log10(p.adjust),size=Count))+
  scale_color_gradient(low="#FF9569",high="#E92758")+
  theme_half_open()+
  labs(x = NULL, y =NULL)+
  theme(axis.text.x = element_text(angle =45))+
  theme(axis.text.x = element_text(vjust = 1,hjust=1))+
  theme(axis.title = element_text(size = 15),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),legend.text = element_text(size = 14))
ggsave("/home/zhzhang/PG/scRNAseq/monocle3/Humantestis.germcellexp_trackDE_modulepglnc_top20coexp.enrich.pdf",
       enrich,width=8,height=8,dpi=1200,units="in",device='pdf',bg ="transparent")


```

