#### FIG 7
##### 1.下载
```r
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR121/034/SRR12164934/SRR12164934_2.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR121/033/SRR12164933/SRR12164933_1.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR121/033/SRR12164933/SRR12164933_2.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR121/036/SRR12164936/SRR12164936_2.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR121/037/SRR12164937/SRR12164937_2.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR121/034/SRR12164934/SRR12164934_1.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR121/038/SRR12164938/SRR12164938_2.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR121/035/SRR12164935/SRR12164935_2.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR121/037/SRR12164937/SRR12164937_1.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR121/035/SRR12164935/SRR12164935_1.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR121/036/SRR12164936/SRR12164936_1.fastq.gz
wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR121/038/SRR12164938/SRR12164938_1.fastq.gz


```
##### 2.cellbarcode矩阵获取
```r
#产生表达矩阵
cd /home/zhzhang/PG/scsper/count/
cat "/home/zhzhang/PG/scsper/sample.name.txt"|while read i
do
arr=($i)
qian=${arr[0]}
hou=${arr[1]}
cellranger count --id ${qian} --transcriptome /home/zhzhang/PG/scRNAseq/crref/grch38_108/ --fastqs /home/zhzhang/PG/scsper/fastq/ --sample ${qian} --nosecondary --force-cells ${hou} --localcores 64 --localmem 2500
done


#样本表达矩阵信息表
find /home/zhzhang/PG/scsper/count/*/filtered_feature_bc_matrix/|awk -F "/" '{print $1"/"$2"/"$3"/"$4"/"$5"/"$6"/"$7"/"$8"/"}'|sort|uniq|awk -F "/" '{print $7"\t"$0}' > /home/zhzhang/PG/scsper/seurat/samplepath.txt



```
##### 3.seurat upstream and sctype cell annotation
```r
input <- "/home/zhzhang/PG/scsper/seurat/samplepath.txt"
output <- "/home/zhzhang/PG/scsper/seurat/integrated_ALLsample.rds"

#生科院服务器，计算每个celltype细胞群中，每个基因的阳性细胞比例pct1
library(dplyr)
library(Seurat)
library(patchwork)
library(future)

plan("multiprocess", workers = 64)
options(future.globals.maxSize= 1250000000000)

#input sample path dataframe
samplepath <- read.delim(input, header=FALSE)
samplenum <- nrow(samplepath)
#Load the dataset
for (i in 1:samplenum) {
  assign(samplepath[i,1],Read10X(data.dir = samplepath[i,2]))
  gc()
}
#创建对象
for (i in 1:samplenum) {
  assign(samplepath[i,1],CreateSeuratObject(counts = get(samplepath[i,1]), project =samplepath[i,1] , min.cells = 10, min.features = 200))
  gc()
}
#细胞质控
#VlnPlot(ALLsample, features = c("nFeature_RNA", "nCount_RNA"), ncol =2, group.by = "orig.ident", pt.size = 0)
for (i in 1:samplenum) {
  assign(samplepath[i,1],PercentageFeatureSet(get(samplepath[i,1]),pattern = "^MT-",col.name = "percent_mt"))
  gc()
}

for (i in 1:samplenum) {
  assign(samplepath[i,1],subset(get(samplepath[i,1]), subset = nFeature_RNA > 500 & percent_mt < 20))
  gc()
}
#标准化归一化
for (i in 1:samplenum) {
  assign(samplepath[i,1],SCTransform(get(samplepath[i,1]),variable.features.n=5000))
  gc()
}

#数据整合（对应新版标准化）
sampleList <- list(C1,C2,C3,N1,N2,N3)
features <- SelectIntegrationFeatures(object.list = sampleList, nfeatures = 5000)
scedata <- PrepSCTIntegration(object.list = sampleList, anchor.features = features)
scedata <- FindIntegrationAnchors(object.list = scedata, normalization.method = "SCT",anchor.features = features,scale=F)
scedata <- IntegrateData(anchorset = scedata, normalization.method = "SCT")

#保存
saveRDS(scedata,file=output)


ALLsample <- readRDS("/home/zhzhang/PG/scsper/seurat/integrated_ALLsample.rds")
#降维
ALLsample <- RunPCA(ALLsample,npcs = 50)
#ElbowPlot(ALLsample, ndims = 50, reduction = "pca")
#聚类
ALLsample <- FindNeighbors(ALLsample, dims = 1:25)
ALLsample <- FindClusters(ALLsample, resolution = 1)
#clustree(ALLsample)
#可视化降维
ALLsample <- RunUMAP(ALLsample, dims = 1:25)
#DimPlot(object = ALLsample, reduction = "umap",dim=c(1,2),group.by="seurat_clusters",label = TRUE)
#RNA
ALLsample <- NormalizeData(ALLsample, assay ="RNA",normalization.method = "LogNormalize", scale.factor = 10000)
ALLsample@active.assay <- "RNA"
saveRDS(ALLsample, file = "/home/zhzhang/PG/scsper/seurat/RNAnormal_ALLsample.rds")
ALLsample.markers <- FindAllMarkers(object = ALLsample,assay="RNA",only.pos = TRUE, min.pct = 0.25, logfc.threshold = 1,return.thresh = 0.01)
data.table::fwrite(ALLsample.markers,file ="/home/zhzhang/PG/scsper/seurat/ALLsample.allcluster.markergene.byRNA.txt",sep = '\t',row.names = F,quote = F,col.names = T)




library(dplyr)
library(Seurat)
library(patchwork)
#导入R
ALLsample <- readRDS("/home/zhzhang/PG/scsper/seurat/RNAnormal_ALLsample.rds")
DimPlot(object = ALLsample, reduction = "umap",group.by="seurat_clusters",label = TRUE)
#手工注释细胞类型
cluster2celltype <- c("0"="RS",
                      "1"="PMC", 
                      "2"="PMC",
                      "3"= "Fibrotic PMC", #fib
                      "4"= "Early SPC", 
                      "5"= "Elongating",
                      "6"= "Elongated", 
                      "7"= "SSC", 
                      "8"= "SSC",
                      "9"= "RS",
                      "10"= "Late SPC",
                      "11"= "RS",
                      "12"= "Late SPC",
                      "13"= "RS",
                      "14"= "Early SPC",
                      "15"= "Late SPC",
                      "16"= "PMC",
                      "17"= "Late SPC",
                      "18"= "Early SPC",
                      "19"= "Late SPC",
                      "20"= "Late SPC",
                      "21"= "Diff.SPG",
                      "22"= "Diff.SPG",
                      "23"= "VSMC",
                      "24"= "Macro",
                      "25"= "EC",
                      "26"= "Sertoli",
                      "27"= "Elongated",
                      "28"= "Sertoli",
                      "29"= "TC",
                      "30"= "Leydig",
                      "31"= "SSC",
                      "32"= "Leydig"
)
ALLsample@meta.data[['cell_type']] <- unname(cluster2celltype[ALLsample@meta.data$seurat_clusters])
#设置每个细胞分类标识的变量名
Idents(ALLsample) <- "cell_type"
#plot
ALLsample@meta.data[['cell_type']] <- factor(ALLsample@meta.data[['cell_type']],
                                             levels =c("SSC","Diff.SPG","Early SPC","Late SPC",
                                                       "RS","Elongating","Elongated",
                                                       "Sertoli","Leydig","PMC","Fibrotic PMC",
                                                       "VSMC","EC",
                                                       "Macro","TC"))
umapcell <- DimPlot(object = ALLsample, reduction = "umap",group.by="cell_type",label = TRUE,label.size=5)+
  scale_color_manual(values = c("#2A6434","#50A15C","#7DBB78","#ACD18E",
                                "#A8CCDC","#317CA4","#2D4D92",
                                "#E3DD93","#984E27","#B86577","#ED7C7E",
                                "#E2A0A4","#B58F7A",
                                "#BFB5D7","#D3A665"))+
  theme(axis.title = element_text(size = 15),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 15))+
  labs(title = NULL,x="UMAP-1",y="UMAP-2")+
  theme(legend.text = element_text(size = 20),legend.key.size = unit(25,"pt"))+
  guides(color=guide_legend(override.aes = list(size=5)))
ggsave("/home/zhzhang/PG/scsper/seurat/Humantestis.celltype_umap.pdf", 
       umapcell,width = 8.5, height = 6.5,dpi=1200, units = "in", device='pdf',bg = "transparent")
#储存
saveRDS(ALLsample, file = "/home/zhzhang/PG/scsper/seurat/celltype_ALLsample.rds")





```
##### 4.celltype三类基因表达比例并获取各cell type表达的基因
```r

input <- "/home/zhzhang/PG/scsper/seurat/celltype_ALLsample.rds"
output <- "/home/zhzhang/PG/scsper/seurat/ALLsample.allcelltype.allgenepct.byRNA.txt"


#生科院服务器脚本，计算每个celltype细胞群中，每个基因的阳性细胞比例pct1
library(dplyr)
library(Seurat)
library(patchwork)
library(future)

ALLsample <- readRDS(input)
plan("multiprocess", workers = 64)
options(future.globals.maxSize= 2500000000000)
ALLsample.markers <- FindAllMarkers(object = ALLsample,assay="RNA",min.pct=0,logfc.threshold=0,only.pos = FALSE,return.thresh=1)

data.table::fwrite(ALLsample.markers,file =output,sep = '\t',row.names = F,quote = F,col.names = T)


```
```r
inputpct <- "/home/zhzhang/PG/scsper/seurat/ALLsample.allcelltype.allgenepct.byRNA.txt"
output1 <- "/home/zhzhang/PG/scsper/plot/allcelltype_3gene_exp_percent.tj.txt"
output2 <- "/home/zhzhang/PG/scsper/plot/allcelltype_3gene_exp_percent.diffpvalue.tj.txt"
output3 <- "/home/zhzhang/PG/scsper/plot/allcelltype_expgene_name_type.txt"


#对比每类细胞中三类基因的表达比例（在某细胞群中阳性细胞比例>=0.01即为在该细胞群中表达）
#导入基因name分类
genename_class <- read.delim("/home/zhzhang/PG/scRNAseq/genename.type.human.txt")
#导入全部基因在各celltype的阳性细胞比例，并添加基因分类
allcelltype_genepct <- read.delim(inputpct)%>%
  select(6,7,3)
colnames(allcelltype_genepct) <- c("celltype","genename","pct.1")
allcelltype_genepct <- left_join(allcelltype_genepct,genename_class,by="genename")
allcelltype_genepct <- allcelltype_genepct[is.na(allcelltype_genepct$type)==F,]
#基因在某细胞群中阳性细胞比例>=0.01即为在该细胞群中表达(后续储存表达的基因列表)
allcelltype_genepct <- mutate(allcelltype_genepct,exp=0)
allcelltype_genepct$exp[allcelltype_genepct$pct.1>=0.01] <- 1
#tj三类基因总数
genenum <- group_by(mutate(genename_class,num=1),type)%>%
  summarise(allnum=sum(num))
#统计三类基因表达比例
celltype_expratio <- group_by(allcelltype_genepct,celltype,type)%>%
  summarise(expnum=sum(exp))%>%
  left_join(genenum,by="type")%>%
  mutate(expratio=expnum/allnum)
celltype_expratio$type <- factor(celltype_expratio$type,levels = c("Protein-coding","Pseudogene-derived lncRNA","Non-pseudogene-derived lncRNA"))

data.table::fwrite(celltype_expratio,file =output1,sep = '\t',row.names = F,quote = F,col.names = T)
#验证人类睾丸每个celltype中pglnc表达比例高于npglnc表达比例的显著性（fishertest）
celltype_expratio <- mutate(celltype_expratio,nonum=allnum-expnum)
celltype_pvalue <- data.frame("celltype"=distinct(celltype_expratio,celltype)$celltype)
for (i in c(1:length(celltype_pvalue$celltype))) {
  fishertest <- fisher.test(filter(celltype_expratio,celltype==celltype_pvalue$celltype[i])[c(1,3),c(3,6)])
  celltype_pvalue[i,2] <- fishertest[["p.value"]]
  if (celltype_pvalue[i,2]>=0.05) {
    celltype_pvalue[i,3] <- "N.S."
  }
  if (celltype_pvalue[i,2]<0.05) {
    celltype_pvalue[i,3] <- "*"
  }
  if (celltype_pvalue[i,2]<0.01) {
    celltype_pvalue[i,3] <- "**"
  }
  if (celltype_pvalue[i,2]<0.001) {
    celltype_pvalue[i,3] <- "***"
  }
}
colnames(celltype_pvalue) <- c("celltype","pvalue","anno")
data.table::fwrite(celltype_pvalue,file =output2,sep = '\t',row.names = F,quote = F,col.names = T)
#输出每个celltype表达的genename及其type
ctexp_genenametype <- filter(allcelltype_genepct,exp==1)%>%
  select(1,2,4)%>%
  mutate(ctagn=paste(celltype,genename,sep = "----"))
data.table::fwrite(ctexp_genenametype,file =output3,sep = '\t',row.names = F,quote = F,col.names = T)
#画图
celltype_expratio$celltype <- factor(celltype_expratio$celltype,
                                     levels = c("SSC","Diff.SPG","Early SPC","Late SPC",
                                                "RS","Elongating","Elongated",
                                                "Sertoli","Leydig","PMC","Fibrotic PMC",
                                                "VSMC","EC",
                                                "Macro","TC"))
ppcell <- ggplot(data = celltype_expratio,aes(x=celltype,y=expratio*100))+
  geom_col(aes(fill=type),width = 0.5,position = "dodge")+
  geom_signif(annotations=celltype_pvalue$anno,y_position=85,tip_length = 0,
              xmin = c(1:15),xmax = c(1.166:15.166))+
  scale_fill_manual(values=c("#BB5A5D","#7197AD","#FAA465","#8491B4"),
                    limits=c("Protein-coding","Pseudogene-derived lncRNA","Non-pseudogene-derived lncRNA"))+
  theme_half_open()+
  scale_y_continuous(limits = c(0,85))+
  labs(x = NULL, y ="Expression proportion (%)",fill = NULL)+
  theme(axis.title = element_text(size = 20),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),
        legend.position = "none")+
  theme(axis.text.x = element_text(angle =30))+
  theme(axis.text.x = element_text(vjust = 1,hjust=1))
ggsave("/home/zhzhang/PG/scsper/plot/Humantestis.lnc_celltype_exp_percent.pdf", 
       ppcell,width = 12, height = 5,dpi=1200, units = "in", device='pdf',bg = "transparent")


```


##### 5.crypto vs normal 每类细胞的差异表达分析
```r

input <- "/home/zhzhang/PG/scsper/seurat/celltype_ALLsample.rds"
output <- "/home/zhzhang/PG/scsper/seurat/ALLsample.allcelltype.CN.DEA.byRNA.txt"
output2 <- "/home/zhzhang/PG/scsper/seurat/ALLsample.allgroup.celltype.cellnum.txt"



#生科院服务器脚本
library(dplyr)
library(Seurat)
library(patchwork)
library(future)

ALLsample <- readRDS(input)
#包括分组的细胞类型注释
ALLsample@meta.data[['group_cell_type']] <- paste(do::Replace(data=ALLsample@meta.data$orig.ident,
                                                              pattern=c("C.*:Crypto","N.*:Normal")),
                                                  ALLsample@meta.data$cell_type,
                                                  sep = "----")
#设置每个细胞分类标识的变量名
Idents(ALLsample) <- "group_cell_type"
#循环对每类细胞，进行O/Y组之间的差异表达分析.
#DEA结果储存在dataframe中并储存
celltype <- distinct(data.frame("celltype"=ALLsample@meta.data[['cell_type']]))%>%filter(celltype!="TC")
celltype <- celltype$celltype
OY <- data.frame()
plan("multiprocess", workers = 64)
options(future.globals.maxSize= 2500000000000)
for (i in c(1:length(celltype))) {
  #OY
  print(paste("now:OY",celltype[i],sep = "----"))
  difexpOY <- FindMarkers(ALLsample,assay="RNA",
                           ident.1=paste("Crypto",celltype[i],sep = "----"),
                           ident.2=paste("Normal",celltype[i],sep = "----"),
                           only.pos = FALSE, min.pct = 0, logfc.threshold =0)%>%
    tibble::rownames_to_column("genename")%>%
    mutate(cell_type=celltype[i])
  OY <- rbind(OY,difexpOY)
}
#储存
data.table::fwrite(OY,
                   file =output,sep = '\t',row.names = F,quote = F,col.names = T)
#细胞频数
pinshu <- data.frame(table(ALLsample@meta.data$group_cell_type))
data.table::fwrite(pinshu,
                   file =output2,sep = '\t',row.names = F,quote = F,col.names = T)




```
##### 每个celltype对比表达的三类基因，在C/N中表达量log2FC和DE比例
```r
input1 <- "/home/zhzhang/PG/scsper/plot/allcelltype_expgene_name_type.txt"
input2 <- "/home/zhzhang/PG/scsper/seurat/ALLsample.allcelltype.CN.DEA.byRNA.txt"
input3 <- "/home/zhzhang/PG/scsper/seurat/ALLsample.allgroup.celltype.cellnum.txt"
output1 <- "/home/zhzhang/PG/scsper/plot/allcelltype_expgene_DEA.txt"
output2 <- "/home/zhzhang/PG/scsper/plot/allcelltype_3gene_DEA.deratio.tj.txt"
output3 <- "/home/zhzhang/PG/scsper/plot/allcelltype_3gene_DEA.deratio.pganpglnc_diffpvalue.tj.txt"


#导入每个celltype表达的genname列表
cte_gnt <- read.delim(input1)%>%
  select(4,3)
#两组比较时，在两组中celltype的num需均满足cellnum > allcellnum*0.01
#导入O/Y testis衰老过程中每类细胞的差异表达分析结果，并添加分类和阳性细胞比例的log2FC
allcelltype_ONY <- read.delim(input2)%>%
  mutate(ctagn=paste(cell_type,genename,sep = "----"))%>%
  left_join(cte_gnt,by="ctagn")%>%
  select(-ctagn)%>%
  mutate(pctlog2FC=log2((pct.1+0.1)/(pct.2+0.1)))
#导入细胞频数
cellnum <- read.delim(input3)
cellnumcutoff <- sum(cellnum$Freq)*0.01
filtercell <- filter(cellnum,Freq<cellnumcutoff)%>%
  separate(Var1,into = c("group","celltype"),sep="----")%>%
  distinct(celltype)
#去除不满足细胞数量的celltype
for (i in 1:length(filtercell$celltype)) {
  allcelltype_ONY <- filter(allcelltype_ONY,cell_type!=filtercell[i,1])
}
#去除不表达的基因
allcelltype_ONY <- allcelltype_ONY[is.na(allcelltype_ONY$type)==F,]
allcelltype_ONY$type <- factor(allcelltype_ONY$type,levels = c("Protein-coding","Pseudogene-derived lncRNA","Non-pseudogene-derived lncRNA"))
#筛选各celltype的ON/Y和ODON差异表达基因(在至少一个细胞群pct>=0.1,|log2FC|>=0.25且padj<0.05 ，de赋为1)
allcelltype_ONYDEG <- mutate(allcelltype_ONY,pctok=0,de=0)
allcelltype_ONYDEG$pctok[which(allcelltype_ONYDEG$pct.1>=0.1 | allcelltype_ONYDEG$pct.2 >=0.1)] <- 1
allcelltype_ONYDEG$de[which(allcelltype_ONYDEG$pctok==1 & allcelltype_ONYDEG$p_val_adj<0.05 & abs(allcelltype_ONYDEG$avg_log2FC)>=0.25)] <- 1
data.table::fwrite(allcelltype_ONYDEG,file =output1,sep = '\t',row.names = F,quote = F,col.names = T)
#统计各celltype，表达的三类基因中，差异表达的比例
tj <- group_by(allcelltype_ONYDEG,cell_type,type)%>%
  summarise(denum=sum(de),num=n())%>%
  mutate(node=num-denum,deratio=denum*100/num)
data.table::fwrite(tj,file =output2,sep = '\t',row.names = F,quote = F,col.names = T)
#验证人类睾丸每个celltype ON/Y中pglnc 差异表达比例高于npglnc的显著性（fisher）
ONY_celltype_pvalue <- data.frame("celltype"=distinct(tj,cell_type)$cell_type)
for (i in c(1:length(ONY_celltype_pvalue$celltype))) {
  test <- fisher.test(filter(tj,cell_type==ONY_celltype_pvalue$celltype[i])[2:3,c(3,5)])
  ONY_celltype_pvalue[i,2] <- test[["p.value"]]
}
colnames(ONY_celltype_pvalue) <- c("celltype","deratiopvalue")
ONY_celltype_pvalue <- mutate(ONY_celltype_pvalue,deratiostr="N.S.")
ONY_celltype_pvalue$deratiostr[ONY_celltype_pvalue$deratiopvalue<0.05] <- "*"
ONY_celltype_pvalue$deratiostr[ONY_celltype_pvalue$deratiopvalue<0.01] <- "**"
ONY_celltype_pvalue$deratiostr[ONY_celltype_pvalue$deratiopvalue<0.001] <- "***"
data.table::fwrite(ONY_celltype_pvalue,file =output3,sep = '\t',row.names = F,quote = F,col.names = T)

#plot 差异表达比例对比
tj$cell_type <- factor(tj$cell_type,levels =c("SSC","Diff.SPG","Early SPC","Late SPC",
                                              "RS","Elongating","Elongated",
                                              "Sertoli","Leydig","PMC","Fibrotic PMC",
                                              "VSMC","EC",
                                              "Macro","TC"))
ONY_celltype_pvalue$celltype <- factor(ONY_celltype_pvalue$celltype,levels =c("SSC","Diff.SPG","Early SPC","Late SPC",
                                                                              "RS","Elongating","Elongated",
                                                                              "Sertoli","Leydig","PMC","Fibrotic PMC",
                                                                              "VSMC","EC",
                                                                              "Macro","TC"))
ONY_celltype_pvalue <- arrange(ONY_celltype_pvalue,celltype)
ONY2 <- ggplot(data = tj,aes(x=cell_type,y=deratio))+
  geom_col(aes(fill=type),width = 0.5,position = "dodge")+
  geom_signif(annotations=ONY_celltype_pvalue$deratiostr,y_position=25,tip_length = 0,
              xmin = c(1:8),xmax = c(1.166:8.166))+
  scale_fill_manual(values=c("#BB5A5D","#7197AD","#FAA465","#8491B4"),
                    limits=c("Protein-coding","Pseudogene-derived lncRNA","Non-pseudogene-derived lncRNA"))+
  theme_half_open()+
  coord_cartesian(ylim = c(0, 25))+
  scale_y_continuous(breaks = c(0,5,10,15,20,25))+
  labs(x = NULL, y ="DE gene proportions (%)",fill = NULL)+
  theme(axis.title = element_text(size = 20),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),
        legend.position = "none")+
  theme(axis.text.x = element_text(angle =30))+
  theme(axis.text.x = element_text(vjust = 1,hjust=1))
ggsave("/home/zhzhang/PG/scsper/plot/Humantestis.celltypeexp_gene_CNderatio.pdf", 
       ONY2,width = 10, height = 5,dpi=1200, units = "in", device='pdf',bg = "transparent")

#2.统计对比差异表达的三类基因 表达量和pct log2fc
allcelltype_ONYDEG$type <- factor(allcelltype_ONYDEG$type,levels = c("Protein-coding","Pseudogene-derived lncRNA","Non-pseudogene-derived lncRNA"))
allcelltype_ONYDEG$cell_type <- factor(allcelltype_ONYDEG$cell_type,
                                    levels = c("SSC","Diff.SPG","Early SPC","Late SPC",
                                               "RS","Elongating","Elongated",
                                               "Sertoli","Leydig","PMC","Fibrotic PMC",
                                               "VSMC","EC",
                                               "Macro","TC"))
#统计
tj <- group_by(allcelltype_ONYDEG,cell_type,type)%>%
  summarise(meanabs_explog2FC=mean(abs(avg_log2FC)),medianabs_explog2FC=median(abs(avg_log2FC)),
            meanabs_pctlog2FC=mean(abs(pctlog2FC)),medianabs_pctlog2FC=median(abs(pctlog2FC)))
data.table::fwrite(tj,file ="/home/zhzhang/PG/scsper/plot/Humantestis.celltypeexp_gene_CNDEA.abslog2FC.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#验证人类睾丸每个celltype O/Y中pglnc expLOG2FC和pctlog2fc绝对值高于npglnc的显著性（wilcox）
ONY_celltype_pvalue <- data.frame("celltype"=c("SSC","Diff.SPG","Early SPC","Late SPC",
                                               "RS","Elongating","Elongated",
                                               "PMC"))
for (i in c(1:nrow(ONY_celltype_pvalue))) {
  test <- wilcox.test(abs(filter(allcelltype_ONYDEG,cell_type==ONY_celltype_pvalue$celltype[i] & type=="Pseudogene-derived lncRNA")$avg_log2FC),
                      abs(filter(allcelltype_ONYDEG,cell_type==ONY_celltype_pvalue$celltype[i] & type=="Non-pseudogene-derived lncRNA")$avg_log2FC))
  ONY_celltype_pvalue[i,2] <- test[["p.value"]]
  test2 <- wilcox.test(abs(filter(allcelltype_ONYDEG,cell_type==ONY_celltype_pvalue$celltype[i] & type=="Pseudogene-derived lncRNA")$pctlog2FC),
                       abs(filter(allcelltype_ONYDEG,cell_type==ONY_celltype_pvalue$celltype[i] & type=="Non-pseudogene-derived lncRNA")$pctlog2FC))
  ONY_celltype_pvalue[i,3] <- test2[["p.value"]]
}
colnames(ONY_celltype_pvalue) <- c("celltype","exppvalue","pctpvalue")
ONY_celltype_pvalue <- mutate(ONY_celltype_pvalue,expstr="N.S.",pctstr="N.S.")
ONY_celltype_pvalue$expstr[ONY_celltype_pvalue$exppvalue<0.05] <- "*"
ONY_celltype_pvalue$expstr[ONY_celltype_pvalue$exppvalue<0.01] <- "**"
ONY_celltype_pvalue$expstr[ONY_celltype_pvalue$exppvalue<0.001] <- "***"
ONY_celltype_pvalue$pctstr[ONY_celltype_pvalue$pctpvalue<0.05] <- "*"
ONY_celltype_pvalue$pctstr[ONY_celltype_pvalue$pctpvalue<0.01] <- "**"
ONY_celltype_pvalue$pctstr[ONY_celltype_pvalue$pctpvalue<0.001] <- "***"
data.table::fwrite(ONY_celltype_pvalue,file ="/home/zhzhang/PG/scsper/plot/Humantestis.celltypeexp_gene_CNabslog2FC.pganpglnc_diffpvalue.tj.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#plot 表达量log2fc对比
ONY <- ggplot(data = allcelltype_ONYDEG,aes(x=cell_type,y=abs(avg_log2FC)))+
  geom_boxplot(aes(fill=type),width = 0.5,outlier.alpha = 0)+
  geom_signif(annotations=ONY_celltype_pvalue$expstr,y_position=1.25,tip_length = 0,
              xmin = c(1:8),xmax = c(1.166:8.166))+
  scale_fill_manual(values=c("#BB5A5D","#7197AD","#FAA465","#8491B4"),
                    limits=c("Protein-coding","Pseudogene-derived lncRNA","Non-pseudogene-derived lncRNA"))+
  theme_half_open()+
  coord_cartesian(ylim = c(0, 1.25))+
  scale_y_continuous(breaks = c(0,0.3,0.6,0.9,1.2))+
  labs(x = NULL, y ="Absolute value\nof expression log2FC",fill = NULL)+
  theme(axis.title = element_text(size = 20),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),
        legend.position = "none")+
  theme(axis.text.x = element_text(angle =30))+
  theme(axis.text.x = element_text(vjust = 1,hjust=1))
ggsave("/home/zhzhang/PG/scsper/plot/Humantestis.celltypeexp_gene_CNabsexplog2FC.pdf", 
       ONY,width = 10, height = 5,dpi=1200, units = "in", device='pdf',bg = "transparent")

```




##### 生殖细胞中，患者DE pglnc的DE coexp基因富集分析，确定DEpglnc的影响
```r
#导入差异表达分析result
allcelltype_expgene_DEA <- read.delim("/home/zhzhang/PG/scsper/plot/allcelltype_expgene_DEA.txt")
#导入pglnc靶基因
pgdlncRNA_transtarget <- read.delim("/home/zhzhang/PG/RNAseq/Homo_sapiens/addtarget_new/Homo_sapiens.spexp.pglnc_pcgene.coexppair.top20.txt")%>%
  select(1,2)
colnames(pgdlncRNA_transtarget)[1] <- "geneid"
#导入ID-name对照表
geneIDname <- read.delim("~/PG/scRNAseq/gene.ID.name.human.txt", header=FALSE)
colnames(geneIDname) <- c("geneid","genename")
geneIDname <- mutate(geneIDname,genename=stringr::str_replace_all(genename,"_","-"))
#pglncid-靶基因id，转变为pglncname-靶基因name
colnames(pgdlncRNA_transtarget) <- c("geneid","trans_acting_target")
pgdlncRNA_transtarget <- left_join(pgdlncRNA_transtarget,geneIDname,by="geneid")%>%
  select(3,2)
colnames(pgdlncRNA_transtarget) <- c("genename","geneid")
pgdlncRNA_transtarget <- left_join(pgdlncRNA_transtarget,geneIDname,by="geneid")%>%
  select(1,3)
colnames(pgdlncRNA_transtarget) <- c("genename","targetname")
#每个生殖细胞循环富集
upenrich1 <- data.frame()
downenrich1 <- data.frame()
celltype <- c("SSC","Diff.SPG","Early SPC","Late SPC","RS","Elongating","Elongated")
for (i in 1:length(celltype)) {
  forcelltype <- celltype[i]
  #提取指定细胞
  sc_DEA <- filter(allcelltype_expgene_DEA,cell_type==forcelltype)
  #提取de蛋白编码基因
  sc_DEA_depcg <- filter(sc_DEA,
                         type=="Protein-coding" & de==1)
  colnames(sc_DEA_depcg) <- paste("target",colnames(sc_DEA_depcg),sep="_")
  colnames(sc_DEA_depcg)[1] <- "targetname"
  #提取depglnc，并获取其对应de靶基因
  sc_DEA_depglnc <- filter(sc_DEA,
                           type=="Pseudogene-derived lncRNA" & de==1)%>%
    left_join(pgdlncRNA_transtarget,by="genename")%>%
    left_join(sc_DEA_depcg,by="targetname")%>%
    mutate(panduan=avg_log2FC*target_avg_log2FC)%>%
    filter(panduan>0)
  #提取上调靶基因
  uptarget <- filter(sc_DEA_depglnc,target_avg_log2FC>0)%>%
    distinct(targetname)
  colnames(uptarget) <- "genename"
  uptarget <- left_join(uptarget,geneIDname,by="genename")%>%
    select(geneid)
  #提取下调靶基因
  downtarget <- filter(sc_DEA_depglnc,target_avg_log2FC<0)%>%
    distinct(targetname)
  colnames(downtarget) <- "genename"
  downtarget <- left_join(downtarget,geneIDname,by="genename")%>%
    select(geneid)
  #GO富集分析
  if (nrow(uptarget)>1) {
    up_tar_GO <- clusterProfiler::enrichGO(gene = uptarget$geneid, OrgDb = 'org.Hs.eg.db', keyType = "ENSEMBL", ont = 'BP', pAdjustMethod = 'fdr', pvalueCutoff = 1, qvalueCutoff = 1)
    up_tar_GO_result <- up_tar_GO@result%>%
      filter(p.adjust<0.05)%>%
      mutate(cell_type=forcelltype)%>%
      mutate(Description=Hmisc::capitalize(Description))%>%
      mutate(Description=do::Replace(data=Description,
                                     pattern=c("NcRNA:ncRNA","MRNA:mRNA",
                                               "RRNA:rRNA","PiRNA:piRNA",
                                               "SnRNA:snRNA")))
    upenrich1 <- rbind(upenrich1,up_tar_GO_result)
  }
  #GO富集分析
  if (nrow(downtarget)>1) {
    down_tar_GO <- clusterProfiler::enrichGO(gene = downtarget$geneid, OrgDb = 'org.Hs.eg.db', keyType = "ENSEMBL", ont = 'BP', pAdjustMethod = 'fdr', pvalueCutoff = 1, qvalueCutoff = 1)
    down_tar_GO_result <- down_tar_GO@result%>%
      filter(p.adjust<0.05)%>%
      mutate(cell_type=forcelltype)%>%
      mutate(Description=Hmisc::capitalize(Description))%>%
      mutate(Description=do::Replace(data=Description,
                                     pattern=c("NcRNA:ncRNA","MRNA:mRNA",
                                               "RRNA:rRNA","PiRNA:piRNA",
                                               "SnRNA:snRNA")))
    downenrich1 <- rbind(downenrich1,down_tar_GO_result)
  }
}

#储存各生殖细胞上调通路
data.table::fwrite(upenrich1,file ="/home/zhzhang/PG/scsper/plot/Humantestis.germcell.up.depglnc_detop20coexp.enrich.txt",sep = '\t',row.names = F,quote = F,col.names = T)
#储存各生殖细胞下调通路
data.table::fwrite(downenrich1,file ="/home/zhzhang/PG/scsper/plot/Humantestis.germcell.down.depglnc_detop20coexp.enrich.txt",sep = '\t',row.names = F,quote = F,col.names = T)

#plot UP
upfilter <- filter(upenrich1,Description!="Cellular process involved in reproduction in multicellular organism")%>%
  filter(cell_type!="RS")%>%
  filter(Count>2)%>%
  group_by(cell_type)%>%
  top_n(7,Count)%>%
  top_n(7,-log10(p.adjust))%>%
  filter(Description!="Regulation of DNA recombination")%>%
  filter(Description!="Positive regulation of DNA metabolic process")
upfilter$cell_type <- factor(upfilter$cell_type,levels = c("SSC","Diff.SPG","Early SPC","Late SPC",
                                                           "RS","Elongating","Elongated"))
up <- ggplot(data = upfilter,aes(x=cell_type,y=Description))+
  geom_point(aes(color=-log10(p.adjust),size=Count))+
  scale_color_gradient(low="#F0DCB4",high="#F9957F",breaks=c(2,4,6,8,10))+
  scale_size(breaks = c(4,8,12),range = c(4,10))+
  theme_half_open()+
  labs(x = NULL, y =NULL)+
  theme(axis.text.x = element_text(angle =90))+
  theme(axis.text.x = element_text(vjust = 0.5,hjust=0.5))+
  theme(axis.title = element_text(size = 15),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),legend.text = element_text(size = 14)) +
  theme(panel.grid.major = element_line(colour = "gray95",size = 0.5),
        panel.grid.minor = element_line(colour = "gray95",size = 0.5)) + 
  theme(panel.grid.major = element_line(linetype = "longdash"),
        panel.grid.minor = element_line(linetype = "longdash"))+
  theme(panel.background = element_rect(colour = "black",size=1))+
  theme(legend.position = "left")

#plot down
downfilter <- filter(downenrich1,Description!="Cellular process involved in reproduction in multicellular organism")%>%
  filter(Description!="Cilium or flagellum-dependent cell motility")%>%
  filter(Count>2)%>%
  group_by(cell_type)%>%
  top_n(5,Count)%>%
  top_n(5,-log10(p.adjust))
downfilter$cell_type <- factor(downfilter$cell_type,levels = c("SSC","Diff.SPG","Early SPC","Late SPC",
                                                               "RS","Elongating","Elongated"))
down <- ggplot(data = downfilter,aes(x=cell_type,y=Description))+
  geom_point(aes(color=-log10(p.adjust),size=Count))+
  scale_color_gradient(low="#29BDD9",high="#276ACE",breaks=c(2,4,6,8,10))+
  scale_size(breaks = c(4,8,12),limits = c(4,15),range = c(4,10))+
  theme_half_open()+
  labs(x = NULL, y =NULL)+
  scale_y_discrete(position = "right")+
  theme(axis.text.x = element_text(angle =90))+
  theme(axis.text.x = element_text(vjust = 0.5,hjust=0.5))+
  theme(axis.title = element_text(size = 15),axis.text.y  = element_text(size = 15),
        axis.text.x = element_text(size = 14),legend.text = element_text(size = 14)) +
  theme(panel.grid.major = element_line(colour = "gray95",size = 0.5),
        panel.grid.minor = element_line(colour = "gray95",size = 0.5)) + 
  theme(panel.grid.major = element_line(linetype = "longdash"),
        panel.grid.minor = element_line(linetype = "longdash"))+
  theme(panel.background = element_rect(colour = "black",size=1))

#合并
pp <- plot_grid(up,down,rel_widths = c(1, 1.3))
ggsave("/home/zhzhang/PG/scsper/plot/Humantestis.germcell.depglnc_detop20coexp.enrich.pdf", 
       pp,width = 16, height = 5,dpi=1200, units = "in", device='pdf',bg = "transparent")



```




